<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A suite of Monte Carlo-Wilcoxon tests • MCWtests</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A suite of Monte Carlo-Wilcoxon tests">
<meta name="description" content="Monte Carlo-Wilcoxon (MCW) tests assess the significance of the difference between two sets of quantitative data by calculating sums of ranks (Wilcoxon) for the dataset under examination before and after rearranging it multiple times (Monte Carlo). This package offers the ability to execute four different MCW tests, each with unique data structures and specific questions to address.">
<meta property="og:description" content="Monte Carlo-Wilcoxon (MCW) tests assess the significance of the difference between two sets of quantitative data by calculating sums of ranks (Wilcoxon) for the dataset under examination before and after rearranging it multiple times (Monte Carlo). This package offers the ability to execute four different MCW tests, each with unique data structures and specific questions to address.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">MCWtests</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/MCWtests.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/diazcastillo/MCWtests/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of MCWtests from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"diazcastillo/MCWtests"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level1">
<div class="page-header"><h1 id="introduction-to-mcw-testing">Introduction to MCW testing<a class="anchor" aria-label="anchor" href="#introduction-to-mcw-testing"></a>
</h1></div>
<p>Originally, Monte Carlo-Wilcoxon (MCW) tests were designed to determined whether the differences between two sets of data were significantly biased in the same direction when compared with what it would be expected by chance. MCW tests proceed by calculating sum-of-ranks-based bias indexes, hence the reference to Frank Wilcoxon who invented the non-parametric rank-sum and signed-rank tests, before and after rearranging the dataset multiple times, hence the Monte Carlo reference often associated to analytical strategies based on repeated random sampling(Díaz-Castillo 2013, 2017, 2018; Chamorro-Garcia et al. 2017; Diaz-Castillo et al. 2019).</p>
<p>The <em>MCWtests</em> package encompasses the original MCW test and three variations that differ in the data structures and the specific questions they interrogate.</p>
<ul>
<li><p>The <strong>matched-measures univariate MCW (muMCW)</strong> <strong>test</strong>, the original MCW test, assesses whether one set of inherently matched-paired measures is significantly biased in the same direction(Díaz-Castillo 2013, 2017, 2018; Chamorro-Garcia et al. 2017; Diaz-Castillo et al. 2019). For instance, muMCW tests can be used to analyze bodyweights or transcript abundances determined at two different timepoints for the same set of mice.</p></li>
<li><p>The <strong>unmatched-measures MCW (uMCW) test</strong> assesses whether two sets of unmatched measures and their heterogeneity are significantly biased in the same direction. For instance, uMCW tests can be used to analyze bodyweights or transcript abundances determined for two sets of mice that have been maintained in different conditions.</p></li>
<li><p>The <strong>matched-measures bivariate MCW (mbMCW) test</strong> assesses whether two sets of inherently matched-paired measures are significantly differentially biased in the same direction. For instance, mbMCW tests can be used to analyze bodyweights or transcript abundances determined at two different timepoints for two sets of mice that have been exposed to different conditions.</p></li>
<li><p>The <strong>bias-measures MCW (bMCW) test</strong> assesses whether a set of measures of bias for a quantitative trait between two conditions or a subset of these bias measures are themselves significantly biased in the same direction. For instance, bMCW tests can be used to analyze bias indexes obtained using other MCW tests or fold change for transcript abundances spanning the entire transcriptome or only for genes located in specific genomic regions from two sets of mice exposed to different conditions.</p></li>
</ul>
</div>
<div class="section level1">
<h1 id="mcw-testing-process">MCW testing process<a class="anchor" aria-label="anchor" href="#mcw-testing-process"></a>
</h1>
<p>Although each MCW test examines distinct data structures to address slightly different questions, all MCW tests share two fundamental steps:</p>
<ol style="list-style-type: decimal">
<li><p><strong>To quantitatively determine the extent and direction of the bias of the measure under analysis</strong>, MCW tests calculate a bias index (BI) by summing ranks and dividing these sums by the maximum possible value of the sums. Consequently, BIs range from 1 to -1 when the measure under analysis is completely biased in each possible direction.</p></li>
<li><p><strong>To determine the significance of the BIs calculated for the user-provided dataset (observed BIs)</strong>, a collection of expected-by-chance BIs is generated by rearranging the original dataset multiple times and calculating BIs for each iteration. <em>P<sub>upper</sub></em> and <em>P<sub>lower</sub></em> values are calculated as the fractions of expected-by-chance BIs that have values higher or equal to and lower or equal to the observed BIs, respectively.</p></li>
</ol>
<p>Each MCW test employs the user-provided parameter <em>max_rearrangements</em> to follow two alternative paths.</p>
<ul>
<li><p><strong>MCW exact tests.</strong> If the number of distinct rearrangements that can be generated from the dataset under analysis is less than <em>max_rearrangements</em>, MCW tests will actually generate all possible data rearrangements to create the collection of expected-by-chance BIs. In this case, <em>P<sub>upper</sub></em> and <em>P<sub>lower</sub></em> values will be exact estimations of the likelihood of obtaining BIs with equal or more extreme values compared to observed BIs with datasets of the same size and range but different internal structures.</p></li>
<li><p><strong>MCW approximated tests.</strong> If the number of distinct rearrangements that can be generated from the dataset under analysis is greater than <em>max_rearrangements</em>, MCW tests will perform a specified number of random data rearrangements, equal to the value of <em>max_rearrangements</em>, to generate the collection of expected-by-chance BIs. In this case, <em>P<sub>upper</sub></em> and <em>P<sub>lower</sub></em> values will represent approximate estimations of the likelihood of obtaining BIs with equal or more extreme values compared to observed BIs with datasets of the same size and range but different internal structures.</p></li>
</ul>
</div>
<div class="section level1">
<h1 id="further-information">Further information<a class="anchor" aria-label="anchor" href="#further-information"></a>
</h1>
<p>Please, see <a href="https://diazcastillo.github.io/MCWtests/"><strong>https://diazcastillo.github.io/MCWtests/</strong></a> for detailed information about the structure of the <em>MCWtests</em> package, including each MCW test and examples.</p>
</div>
<div class="section level1">
<h1 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h1>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-chamorro-garcia2017" class="csl-entry">
Chamorro-Garcia, Raquel, Carlos Diaz-Castillo, Bassem M. Shoucri, Heidi Käch, Ron Leavitt, Toshi Shioda, and Bruce Blumberg. 2017. “Ancestral Perinatal Obesogen Exposure Results in a Transgenerational Thrifty Phenotype in Mice.” <em>Nature Communications</em> 8 (1). <a href="https://doi.org/10.1038/s41467-017-01944-z" class="external-link uri">https://doi.org/10.1038/s41467-017-01944-z</a>.
</div>
<div id="ref-diaz-castillo2019" class="csl-entry">
Diaz-Castillo, Carlos, Raquel Chamorro-Garcia, Toshi Shioda, and Bruce Blumberg. 2019. “Transgenerational Self-Reconstruction of Disrupted Chromatin Organization After Exposure To An Environmental Stressor in Mice.” <em>Scientific Reports</em> 9 (1). <a href="https://doi.org/10.1038/s41598-019-49440-2" class="external-link uri">https://doi.org/10.1038/s41598-019-49440-2</a>.
</div>
<div id="ref-díaz-castillo2013" class="csl-entry">
Díaz-Castillo, Carlos. 2013. “Females and Males Contribute in Opposite Ways to the Evolution of Gene Order in Drosophila.” Edited by Juan F. Poyatos. <em>PLoS ONE</em> 8 (5): e64491. <a href="https://doi.org/10.1371/journal.pone.0064491" class="external-link uri">https://doi.org/10.1371/journal.pone.0064491</a>.
</div>
<div id="ref-díaz-castillo2017" class="csl-entry">
———. 2017. “Transcriptome Dynamics Along Axolotl Regenerative Development Are Consistent with an Extensive Reduction in Gene Expression Heterogeneity in Dedifferentiated Cells.” <em>PeerJ</em> 5 (November): e4004. <a href="https://doi.org/10.7717/peerj.4004" class="external-link uri">https://doi.org/10.7717/peerj.4004</a>.
</div>
<div id="ref-díaz-castillo2018" class="csl-entry">
———. 2018. “Same-Sex Twin Pair Phenotypic Correlations Are Consistent with Human Y Chromosome Promoting Phenotypic Heterogeneity.” <em>Evolutionary Biology</em> 45 (3): 248–58. <a href="https://doi.org/10.1007/s11692-018-9454-y" class="external-link uri">https://doi.org/10.1007/s11692-018-9454-y</a>.
</div>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/diazcastillo/MCWtests/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/diazcastillo/MCWtests/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3) + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing MCWtests</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Carlos Diaz-Castillo <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-5099-3619" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Raquel Chamorro-Garcia <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-9235-9499" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Carlos Diaz-Castillo, Raquel Chamorro-Garcia.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
