---
title: "MCWtests"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MCWtests}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: references.bib
csl: nature.csl
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "",
  dev = "svg",
  warning = FALSE,
  message = FALSE
)
```

```{r setup, echo = FALSE}
library(MCWtests)
library(data.table)
library(DiagrammeR)
library(ggplot2)
```

# 1. Introduction to MCW testing

Originally, Monte Carlo-Wilcoxon (MCW) tests were designed to determined whether the differences between two sets of data were significantly biased in the same direction when compared with what it would be expected by chance. MCW tests proceed by calculating sum-of-ranks-based bias indexes, hence the reference to Frank Wilcoxon who invented the non-parametric rank-sum and signed-rank tests, before and after rearranging the dataset multiple times, hence the Monte Carlo reference often associated to analytical strategies based on repeated random sampling [@díaz-castillo2013; @díaz-castillo2017; @chamorro-garcia2017; @díaz-castillo2018; @diaz-castillo2019].

The *MCWtests* package encompasses the original MCW test and three variations that differ in the data structures and the specific questions they interrogate.

-   The **matched-measures univariate MCW (muMCW)** **test**, the original MCW test[@díaz-castillo2013; @díaz-castillo2017; @chamorro-garcia2017; @díaz-castillo2018; @diaz-castillo2019], assesses whether one set of inherently matched-paired measures is significantly biased in the same direction. For instance, muMCW tests can be used to analyze bodyweights or transcript abundances determined at two different timepoints for the same set of mice.

-   The **unmatched-measures MCW (uMCW) test** assesses whether two sets of unmatched measures and their heterogeneity are significantly biased in the same direction. For instance, uMCW tests can be used to analyze bodyweights or transcript abundances determined for two sets of mice that have been maintained in different conditions.

-   The **matched-measures bivariate MCW (mbMCW) test** assesses whether two sets of inherently matched-paired measures are significantly differentially biased in the same direction. For instance, mbMCW tests can be used to analyze bodyweights or transcript abundances determined at two different timepoints for two sets of mice that have been exposed to different conditions.

-   The **bias-measures MCW (bMCW) test** assesses whether a set of measures of bias for a quantitative trait between two conditions or a subset of these bias measures are themselves significantly biased in the same direction. For instance, bMCW tests can be used to analyze bias indexes obtained using other MCW tests or fold change for transcript abundances spanning the entire transcriptome or only for genes located in specific genomic regions from two sets of mice exposed to different conditions.

# 2. MCW testing process

Although each MCW test examines distinct data structures to address slightly different questions, all MCW tests share two fundamental steps:

1.  **To quantitatively determine the extent and direction of the bias of the measure under analysis**, MCW tests calculate a bias index (BI) by summing ranks and dividing these sums by the maximum possible value of the sums. Consequently, BIs range from 1 to -1 when the measure under analysis is completely biased in each possible direction.

2.  **To determine the significance of the BIs calculated for the user-provided dataset (observed BIs)**, a collection of expected-by-chance BIs is generated by rearranging the original dataset multiple times and calculating BIs for each iteration. *P~upper~* and *P~lower~* values are calculated as the fractions of expected-by-chance BIs that have values higher or equal to and lower or equal to the observed BIs, respectively.

Each MCW test employs the user-provided parameter *max_rearrangements* to follow two alternative paths.

```         
-   **MCW exact tests.** If the number of distinct rearrangements that can be generated from the dataset under analysis is less than *max_rearrangements*, MCW tests will actually generate all possible data rearrangements to create the collection of expected-by-chance BIs. In this case, *P~upper~* and *P~lower~* values will be exact estimations of the likelihood of obtaining BIs with equal or more extreme values compared to observed BIs with datasets of the same size and range but different internal structures.

-   **MCW approximated tests.** If the number of distinct rearrangements that can be generated from the dataset under analysis is greater than *max_rearrangements*, MCW tests will perform a specified number of random data rearrangements, equal to the value of *max_rearrangements*, to generate the collection of expected-by-chance BIs. In this case, *P~upper~* and *P~lower~* values will represent approximate estimations of the likelihood of obtaining BIs with equal or more extreme values compared to observed BIs with datasets of the same size and range but different internal structures.
```

# 3. MCWtests data and function structure

Each MCW test involves two local files, one exported function that users interacts with, and several internal functions that perform specific tasks.

```{r, echo = FALSE, out.width="100%"}
DiagrammeR::grViz("digraph flowchart {
                  
                  entry_dataset1 [shape = folder, label = \"entry csv dataset\", color = lightblue, style = filled]
                  results [shape = folder, label = \"results csv table\", color = yellowgreen, style = filled]
                  function1 [shape = box, label = \"MCWtest\", color = salmon, style = filled]
                  function2 [shape = box, label = \"MCW_exact_test\", color = grey, style = filled]
                  function3 [shape = box, label = \"MCW_approximated_test\", color = grey, style = filled]
                  function4 [shape = box, label = \"MCW_exact_test_combinations\", color = grey, style = filled]
                  function5 [shape = box, label = \"MCW_approximated_test_observed_BIs\", color = grey, style = filled]
                  function6 [shape = box, label = \"MCW_approximated_test_simulated_BIs\", color = grey, style = filled]
                  
                  entry_dataset1 -> function1
                  function1 -> results
                  function1 -> function2
                  function1 -> function2 [dir='back']
                  function1 -> function3
                  function1 -> function3 [dir='back']
                  function2 -> function4
                  function2 -> function4 [dir='back']
                  function3 -> function5
                  function3 -> function5 [dir='back']
                  function3 -> function6
                  function3 -> function6 [dir='back']
                  function5 -> function6
                  function5 -> function6 [dir='back']
                
                  {rank = same; entry_dataset1 results}
                  {rank = same; function4 function5 function6}
                  
                  }")
```

1.  **Local files:**

-   **Entry CSV dataset**. Users provide data for MCW testing in a CSV file in the directory of their choice. This file can contain data for a single MCW test or multiple MCW tests to be executed simultaneously. Users can organize the entry dataset in two different layouts, vertical or horizontal, which is particularly useful for simultaneous testing when the data structures for each individual test are significantly different or very similar, respectively.

-   **Results CSV table**. MCW test functions create a CSV file with the results of MCW testing, with a similar name to the entry CSV dataset file, in the same directory where the entry CSV is located.

2.  **Exported functions:**

-   ***MCWtest*** **function**: When executing this function, users must specify the path to the CSV file containing the entry dataset and the parameter *max_rearrangements*. The *MCWtest* function interprets this parameter as the maximum number of data rearrangement iterations used to define the size of the collection of expected-by-chance BIs. The *MCWtest* function performs the following tasks:

    -   It loads the entry CSV dataset.
    -   It determines whether MCW tests will proceed using the exact or approximated testing paths.
    -   It calls the corresponding functions, *MCW_exact_test* and *MCW_approximated_test*, to run the exact and approximated tests, resepctively.
    -   It writes the results of MCW testing to a CSV file in the same directory as the entry CSV file.

3.  **Internal functions:**

-   ***MCW_exact_test*** **function**: This function is called when the number of distinctive data rearrangements is less than the user-provided parameter *max_rearrangements*. The *MCW_exact_test* function performs the following tasks:

    -   It generates all distinctive data rearrangements, either on its own or by executing the *MCW_exact_test_combinations* function.
    -   It calculates BIs for the user-provided data and each of the distinctive rearrangements of the dataset.
    -   It calculates the *P~upper~* and *P~lower~* values as the fractions of expected-by-chance BIs with values higher or equal to and lower or equal to the observed BIs, respectively.
    -   It returns the MCW exact testing results to the *MCWtest* function.

-   ***MCW_exact_test_combinations*** **function**: This function is called to assist the *MCW_exact_test* function in generating all distinctive data rearrangements.

-   ***MCW_approximated_test*** **function**: This function is called when the number of distinctive data rearrangements is greater than the user-provided parameter *max_rearrangements*. The *MCW_approximated_test* function performs the following tasks:

    -   It calls the *MCW_approximated_test_observed_BIs* function to calculate BIs for the user-provided dataset.
    -   It calls the *MCW_approximated_test_simulated_BIs* function to randomly rearrange the data and calculate the expected-by-chance BIs a specified number of times, which is equal to *max_rearrangements*.
    -   It calculates the *P~upper~* and *P~lower~* values as the fractions of expected-by-chance BIs with values higher or equal to and lower or equal to the observed BIs, respectively.
    -   It returns the MCW approximated testing results to the *MCWtest* function

-   ***MCW_approximated_test_observed_BIs*** **function**: This function is called to calculate BIs for the user-provided dataset.

-   ***MCW_approximated_test_simulated_BIs*** **function**: This function is called N times, where N = *max_rearrangements*, to randomly rearrange the data and calculate the expected-by-chance BIs using the *MCW_approximated_test_observed_BIs* function.

# 4. Specific MCW tests

## 4.1 unmatched-measures MCW (uMCW) test

### Introduction

The uMCW test is actually a combination of two different tests that assess whether two sets of unmatched measures and their heterogeneity are significantly biased in the same direction. Significantly different data heterogeneities between two conditions could indicate that the measure under analysis is more constrained or more relaxed in one of the conditions, potentially providing insights into the mechanisms underlying the variation of such measure. For instance, uMCW tests can be used to analyze bodyweights or transcript abundances determined for two sets of mice that have been maintained in different conditions.

### uMCW testing entry dataset formatting

When executing the *uMCWtest* function, users must provide the path to a local CSV file named *X_uMCWtest_data.csv*, where *X* serves as a user-defined identifier. *X_uMCWtest_data.csv* can be structured in two distinct formats:

-   **Vertical layout**: This format allows appending datasets with varying structures, such as different numbers of measures per set or between each appended test. Vertical entry datasets should include the following columns:

    -   The *condition* column uniquely identifies each of the two measure sets under analysis.
    -   The *value* column contains the actual measures under analysis.
    -   As many informative columns as needed by users to contextualize the results of each test. The names of these columns should not include the terms *condition* or *value*. While these columns are optional when running a single test, at least one column is required when running multiple tests simultaneously. All rows for each individual test must contain the same information in these columns.

-   **Horizontal layout**: This format allows appending datasets with similar structures, such as the same number of measures collected for each of the two conditions. Horizontal entry datasets should include the following columns:

    -   Columns *condition_a* and *condition_b* uniquely identify the two measure sets under analysis.
    -   Columns *a.i* and *b.j*, where *i* and *j* represent integers to differentiate specific measures within each set, contain the actual measures under analysis.
    -   As many informative columns as needed by users to contextualize the results of each test. The names of these columns should not contain the term *condition* or have the same structure as the *a.i* and *b.j* columns. While these columns are optional when running a single test, at least one column is required when running multiple tests simultaneously.

### uMCW testing process

The function *uMCWtest* eliminate missing values (NAs) from the dataset before proceeding these steps.

-   To estimate the bias between the two sets of measures (*e.g.*, *a* and *b*), the function *uMCWtest* performs these tasks:

    -   It generates all possible disjoint data pairs using measures from both sets.
    -   For each measure pair, it subtracts the second measure in the pair from the first measure in the pair.
    -   It ranks the absolute values of all non-zero measure pair differences from lowest to highest. Measure pair differences with a value of 0 are assigned a 0 rank. If multiple measure pair differences have the same absolute value, all tied measure pair differences are assigned the lowest rank possible.
    -   It assigns each measure pair rank a sign based on the sign of its corresponding measure pair difference.
    -   It sums the signed ranks for measure pairs formed with measures from the two different sets (*e.g.*, *a-b* and *b-a*).
    -   For each type of disjoint set measure pairs (*e.g.*, *a-b* and *b-a*), it calculates uMCW_BI by dividing the sum of signed ranks by the maximum number this sum could have if the corresponding measure pairs had the highest possible positive ranks. Consequently, uMCW_BI ranges between 1 when all the values for measures in the first set are higher than all the values from measures in the second set, and -1 when all the values for measures in the first set are lower than all the values from measures in the second set.

-   To estimate the bias between the heterogeneity of two sets of measures, the function *uMCWtest* performs these tasks:

    -   It generates all possible disjoint data pairs within each set, disregarding the order of the paired measures. For instance, the measure pair *a.1-a.2* is considered equivalent to the measure pair *a.2-a.1*, and only the former is retained for the subsequent calculations.
    -   For each measure pair, it subtracts the second measure from the first measure.
    -   It ranks all measure pair differences with non-zero values from lowest to highest. Measure pair differences with a value of 0 are assigned a 0 rank. If multiple measure pair differences have the same absolute value, *uMCWtest* assigns all tied measure pair differences the lowest rank possible.
    -   It sums ranks for measure pairs formed with measures from the same set (*e.g.*, *a-a* and *b-b*).
    -   For each type of same-set measure pairs (*e.g.*, *a-a* and *b-b*), it divides each sum of signed ranks by the maximum number this sum could have if the corresponding measure pairs had the highest possible ranks.
    -   It calculates two heterogeneity bias indexes (uMCW_HBIs) by subtracting the normalized sum of signed ranks from the previous step in two possible directions (*e.g.*, *a-b* and *b-a*). Consequently, uMCW_HBI ranges between 1 when at least two measures in the first set have distinct values and all measures in the second set have the same value, and -1 when all measures in the first set have the same value and at least two measures in the second set have distinct values.

-   To assess the significance of the uMCW_BIs and uMCW_HBIs obtained with the user-provided data (observed uMCW_BIs and uMCW_HBIs), the function *uMCWtest* performs these tasks:

    -   It generates a collection of expected-by-chance uMCW_BIs and uMCW_HBIs. These expected values are obtained by rearranging the measures between the two sets multiple times. The user-provided parameter *max_rearrangements* determines the two paths that the function *uMCWtest* can follow to generate the collection of expected-by-chance uMCW_BIs and uMCW_HBIs:

        -   *uMCW exact testing*: If the number of distinct measure rearrangements that can alter their initial set distribution is less than *max_rearrangements*, the function *uMCWtest* calculates uMCW_BIs and uMCW_HBIs for all possible data rearrangements.
        -   *uMCW approximated testing*: If the number of distinct measure rearrangements that can alter their initial set distribution is greater than *max_rearrangements*, the function *uMCWtest* will perform N = *max_rearrangements* random measure rearrangements to calculate the collection of expected-by-chance uMCW_BIs and uMCW_HBIs.

    -   It calculates *P~upper~* and *P~lower~* values as the fraction of expected-by-chance uMCW_BIs and uMCW_HBIs that are higher or equal to and lower or equal to the observed uMCW_BIs and uMCW_HBIs, respectively.

### uMCW testing results

The *uMCWtest* function reports to the console the total number of tests it will execute, and their exact and approximated counts. It also creates a CSV file named *X_uMCWtest_results.csv*, where *X* is a user-defined identifier for the entry dataset CSV file. The *X_uMCWtest_results.csv* file contains four rows for each uMCWtest, two for uMCW_BIs calculated for each condition contrast (*e.g.*, *a-b* and *b-a*), and two for uMCW_HBIs calculated for each condition contrast.

The *X_uMCWtest_results.csv* file includes the following columns:

-   User-provided informative columns to contextualize the results of each test.
-   Columns *condition_a* and *condition_b* indicate the two measure sets under analysis.
-   Columns *N*, *n_a* and *n_b* indicate the total number of measures and the number of measures belonging to each set after removing missing values (NAs).
-   Column *test_type* distinguishes between exact and approximated tests.
-   Column *BI_type* indicates the bias index type (uMCW_BI and uMCW_HBI) for each row of results.
-   Column *condition_contrast* indicates the set contrast (*e.g.*, *a-b* or *b-a*) for each row of results.
-   Column *observed_BI* contains the values of uMCW_BIs and uMCW_HBIs obtained from analyzing the user-provided dataset.
-   Column *expected_by_chance_BI_N* indicates the number of data rearrangements used to calculate the expected-by-chance uMCW_BIs and uMCW_HBIs. This value corresponds to the lowest number between all possible measure rearrangements and the parameter *max_rearrangements*.
-   Columns *pupper* and *plower* represent the *P~upper~* and *P~lower~* values, respectively. They denote the fraction of expected-by-chance uMCW_BIs or uMCW_HBIs with values higher or equal to and lower or equal to the observed uMCW_BIs or uMCW_HBIs, respectively.

### Examples

#### uMCW testing of multiple datasets organized vertically

This example of uMCW testing entry dataset with a vertical layout showcases data for three ideal tests (Table 1). The *Contrast* I test includes data for two sets of measures randomly selected from the same range (Figure 1). The *Contrast* II test includes data representing two sets of measures randomly selected from two different ranges, with measures from the second set being generally higher than those from the first set (Figure 1). The *Contrast* III test includes data representing two sets of measures randomly selected from two different ranges, with measures from the second set being generally higher and less heterogeneous than those from the first set (Figure 1). The results of these uMCW tests align with the intended structure of the entry datasets (Table 2).

##### Table 1. uMCW testing entry dataset with vertical layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_vertical_uMCWtest_data.csv", package = "MCWtests"))
```

##### Figure 1. Data structure of uMCW testing entry dataset with vertical layout

```{r, echo = FALSE, fig.width = 6, fig.height = 5, fig.align = 'center'}
ggplot (data.table::fread(system.file("extdata", "example_vertical_uMCWtest_data.csv", package = "MCWtests")), aes(x = condition, y = value, fill = condition, colour = condition)) +
  geom_dotplot(binaxis = "y", stackdir = "center", dotsize = 0.7) +
  facet_grid (. ~ contrast) +
  theme (legend.title = element_text(size = 10),
         legend.text = element_text(size = 10),
         legend.key.size = unit(0.2, "cm"),
         axis.text.x = element_text(size = 10))
```

##### Table 2. Results of uMCW testing using an entry dataset with vertical layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_vertical_uMCWtest_results.csv", package = "MCWtests"))
```

#### uMCW testing of multiple datasets organized horizontally

This example of uMCW testing entry dataset with a horizontal layout showcases data for nine ideal tests (Table 3). It includes two types of informative columns to accommodate cases where users might want to perform multiple tests with a nested structure. The columns with the *contrast* prefix provide contextual information for groups of tests or rows, while the columns with the *element* prefix provide contextual information for each test or row. Each *Contrast* I row includes two sets of measures randomly selected from the same range (Figure 2). Each *Contrast* II row includes two sets of measures randomly selected from two different ranges, with the measures from the second set being generally higher than those from the first set (Figure 2). Each *Contrast* III row includes two sets of measures randomly selected from two different ranges, with the measures from the second set being generally higher and less heterogeneous than those from the first set (Figure 1). The results of these uMCW tests align with the intended structure of the entry datasets (Table 4).

##### Table 3. uMCW testing entry dataset with horizontal layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_horizontal_uMCWtest_data.csv", package = "MCWtests"))
```

##### Figure 2. Data structure of uMCW testing entry dataset with horizontal layout

```{r, echo = FALSE, fig.width = 6, fig.height = 5, fig.align = 'center'}
ggplot (na.omit(melt(fread(system.file("extdata", "example_horizontal_uMCWtest_data.csv", package = "MCWtests")),
                                     measure.vars = patterns("\\."),
                                     variable.name = "condition",
                                     value.name = "value",
                                     value.factor = FALSE)[
                                       , `:=`(c("condition", "replicate"), tstrsplit(condition, ".", fixed = TRUE))][
                                         , `:=`(replicate, NULL)])[
                                           , `:=`(condition = fifelse(condition == "a", "AAAA", "BBBB"))],
        aes(x = element_ID, y = value, fill = condition, colour = condition)) +
  geom_dotplot(binaxis = "y", stackdir = "center", position = "dodge", dotsize = 0.4) +
  facet_grid (. ~ contrast) +
  theme (legend.title = element_text(size = 10),
         legend.text = element_text(size = 10),
         legend.key.size = unit(0.2, "cm"),
         axis.text.x = element_text(size = 10)) +
  labs(x = "element_ID") 
```

##### Table 4. Results of uMCW testing using an entry dataset with horizontal layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_horizontal_uMCWtest_results.csv", package = "MCWtests"))
```

## 4.2 matched-measures univariate MCW (muMCW) test

### Introduction

The muMCW test assesses whether one set of inherently matched-paired measures is significantly biased in the same direction. For instance, muMCW tests can be used to analyze bodyweights or transcript abundances determined at two different timepoints for the same set of mice.

### muMCW testing entry dataset formatting

When executing the *muMCWtest* function, users must provide the path to a local CSV file named *X_muMCWtest_data.csv*, where *X* serves as a user-defined identifier. *X_muMCWtest_data.csv* can be structured in two distinct formats:

-   **Vertical layout:** This format allows appending datasets with varying structures, such as different numbers of measure matched-pairs for each appended test. Vertical entry datasets should include the following columns:

    -   Columns *condition_a* and *condition_b* uniquely identify the two conditions under which matched-paired measures were collected.
    -   Columns *value_a* and *value_b* contain the actual measures under analysis.
    -   As many informative columns as needed by users to contextualize the results of each test. The names of the these columns should not contain the terms *condition* or *value*. While these columns are optional when running a single test, at least one column is required when running multiple tests simultaneously. All rows for each individual test must contain the same information in these columns.

-   **Horizontal layout:** This format allows appending datasets with similar structures, such as the same number of matched-paired measures for each appended test. Horizontal entry datasets should include the following columns:

    -   Columns *condition_a* and *condition_b* uniquely identify the two conditions under which matched-paired measures were collected.
    -   Columns *a.i* and *b.i*, where *i* represents integers to differentiate each specific matched-pairs of measures, contain the actual measures under analysis.
    -   As many informative columns as needed by users to contextualize the results of each test. The names of these columns should not contain the term *condition* or have the same structure as the *a.i* and *b.i* columns. While these columns are optional when running a single test, at least one column is required when running multiple tests simultaneously.

### muMCW testing process

The function *muMCWtest* eliminates any matched-paired measures with at least one missing value (NA) before proceeding with the following steps.

-   To estimate the bias for all matched-paired measures in the dataset, the function *muMCWtest* performs the following tasks:

    -   For each matched-pair of measures, it subtracts values for the two possible condition contrasts (*e.g.*, *a-b* and *b-a*).
    -   For each condition contrast, it ranks the absolute values of non-zero differences from lowest to highest. Measure pair differences with a value of 0 are assigned a 0 rank. If multiple measure pair differences have the same absolute value, all tied measure pair differences are assigned the lowest rank possible.
    -   It assigns each measure pair rank a sign based on the sign of its corresponding measure pair difference.
    -   It sums the signed ranks for each condition contrast.
    -   It calculates muMCW_BI by dividing each sum of signed ranks by the maximum number that sum could have if the corresponding measure pairs had the highest possible positive ranks. Consequently, muMCW_BI ranges between 1 when all measures corresponding to the first condition are higher than all measures corresponding to the second condition, and -1 when all measures corresponding to the first condition are lower than all measures corresponding to the second condition.

-   To assess the significance of the muMCW_BIs obtained from the user-provided dataset (observed muMCW_BIs), the function *muMCWtest* performs the following tasks:

    -   It generates a collection of expected-by-chance muMCW_BIs. These expected values are obtained by rearranging the measures between and within the two conditions multiple times. The user-provided parameter *max_rearrangements* determines the two paths that the function *muMCWtest* can follow to generate the collection of expected-by-chance muMCW_BIs:

        -   *muMCW exact testing*: If the number of distinct measure rearrangements that can alter their initial pair and set distribution is less than *max_rearrangements*, the function *muMCWtest* calculates muMCW_BIs for all possible data rearrangements.
        -   *muMCW approximated testing*: If the number of distinct measure rearrangements that can alter their initial pair and set distribution is greater than *max_rearrangements*, the function *muMCWtest* performs N = *max_rearrangements* random measure rearrangements to calculate the collection of expected-by-chance muMCW_BIs.

    -   It calculates the *P~upper~* and *P~lower~* values as the fraction of expected-by-chance muMCW_BIs that are higher or equal to and lower or equal to the observed muMCW_BIs, respectively.

### muMCW testing results

The *muMCWtest* function reports to the console the total number of tests it will execute, and their exact and approximated counts. It also creates a CSV file named *X_muMCWtest_results.csv* where *X* is a user-defined identifier for the entry dataset CSV file. The *X_muMCWtest_results.csv* file contains two rows for each muMCW test, with muMCW_BIs calculated for each possible condition contrast (*e.g.*, *a-b* and *b-a*).

The *X_muMCWtest_results.csv* file includes the following columns:

-   User-provided informative columns to contextualize the results of each test.
-   Columns *condition_a* and *condition_b* indicate the two conditions for which matched-paired measures were provided.
-   Column *N* indicates the total number of measure matched-pairs after removing matched-pairs with missing values (NAs).
-   Column *test_type* distinguishes between exact and approximated tests.
-   Column *BI_type* indicates muMCW_BI.
-   Column *condition_contrast* indicates the condition contrast for each row of results.
-   Column *observed_BI* contains the value of muMCW_BIs obtained from analyzing the user-provided dataset.
-   Column *expected_by_chance_BI_N* indicates the number of data rearrangements used to calculate the expected-by-chance muMCW_BIs. This value corresponds to the lowest number between all possible measure rearrangements and the parameter *max_rearrangements*.
-   Columns *pupper* and *plower* represent *P~upper~* and *P~lower~* values, respectively. They denote the fraction of expected-by-chance muMCW_BIs with values higher or equal to and lower or equal to the observed muMCW_BIs, respectively.

### Examples

#### muMCW testing of multiple datasets organized vertically

This example of muMCW testing entry dataset with a vertical layout showcases data for two ideal tests (Table 5). The *Contrast* I test includes data representing one set of matched-paired measures randomly selected from the same range (Figure 3). The *Contrast* II test includes data representing one set of matched-paired measures randomly selected from two different ranges, with measures from the second condition in each matched-pair being generally higher than measures from the first condition in each matched-pair (Figure 3). The results of these muMCW tests align with the intended structure of the entry datasets (Table 6).

##### Table 5. muMCW testing entry dataset with vertical layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_vertical_muMCWtest_data.csv", package = "MCWtests"))
```

##### Figure 3. Data structure of muMCW testing entry dataset with vertical layout

```{r, echo = FALSE, fig.width = 6, fig.height = 5, fig.align = 'center'}
ggplot(melt(na.omit(data.table::fread(system.file("extdata", "example_vertical_muMCWtest_data.csv", package = "MCWtests")))[
  , `:=`(replicate_pair, seq_len(.N)), by = contrast],
  id.vars = c("contrast", "replicate_pair"),
  measure.vars = patterns("condition", "value"),
  variable.name = "pair_element",
  value.name = c("condition", "value")), 
  aes(x = condition, y = value, fill = condition)) +
  geom_dotplot(binaxis = "y", stackdir = "center", aes(colour = condition)) +
  geom_line(linewidth = 0.2, colour = "black", aes(group = replicate_pair)) +
  facet_grid(. ~ contrast) +
  theme (legend.title = element_text(size = 10),
         legend.text = element_text(size = 10),
         legend.key.size = unit(0.2, "cm"),
         axis.text.x = element_text(size = 10))
```

##### Table 6. Results of muMCW testing using an entry dataset with vertical layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_vertical_muMCWtest_results.csv", package = "MCWtests"))
```

#### muMCW testing of multiple datasets organized horizontally

This example of muMCW testing entry dataset with a horizontal layout showcases data for six ideal tests (Table 7). It includes two types of informative columns to accommodate cases where users might want to perform multiple tests with a nested structure. The columns with the *contrast* prefix provide contextual information for groups of tests or rows, while the columns with the *element* prefix provide contextual information for each test or row. Each *Contrast* I row includes one set of matched-paired measures randomly selected from the same range (Figure 4). Each *Contrast* II row includes one set of matched-paired measures randomly selected from two different ranges, with the measures for the second condition in each matched-pair being generally higher than measures from the first condition in each matched-pair (Figure 4). The results of these muMCW tests align with the intended structure of the entry datasets (Table 8).

##### Table 7. muMCW testing entry dataset with horizontal layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_horizontal_muMCWtest_data.csv", package = "MCWtests"))
```

##### Figure 4. Data structure of muMCW testing entry dataset with horizontal layout

```{r, echo = FALSE, fig.width = 6, fig.height = 5, fig.align = 'center'}
ggplot(melt(na.omit(dcast(melt(data.table::fread(system.file("extdata", "example_horizontal_muMCWtest_data.csv", package = "MCWtests")),
                                            measure.vars = patterns("\\."),
                                            variable.name = "condition",
                                            value.name = "value",
                                            value.factor = FALSE)[
                                              , `:=`(c("condition", "replicate_pair"), tstrsplit(condition, ".", fixed = TRUE))][
                                                , `:=`(replicate_pair = as.numeric(replicate_pair))][
                                                  , `:=`(condition, paste0("value_", condition))],
                                       ... ~ condition, value.var = "value")),
          id.vars = c("contrast", "element_ID", "replicate_pair"),
  measure.vars = patterns("condition", "value"),
  variable.name = "pair_element",
  value.name = c("condition", "value")), 
  aes(x = condition, y = value, fill = condition, color = condition)) +
  geom_dotplot(binaxis = "y", stackdir = "center", position = "dodge", aes(colour = condition)) +
  geom_line(linewidth = 0.2, colour = "black", aes(group = replicate_pair)) +
  facet_grid(contrast ~ element_ID, scales = "free") +
  theme (legend.title = element_text(size = 10),
         legend.text = element_text(size = 10),
         legend.key.size = unit(0.2, "cm"),
         axis.text.x = element_text(size = 10))
```

##### Table 8. Results of muMCW testing using an entry dataset with horizontal layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_horizontal_muMCWtest_results.csv", package = "MCWtests"))
```

## 4.3 matched-measures bivariate MCW (mbMCW) test

### Introduction

The mbMCW test assesses whether two sets of inherently matched-paired measures are significantly differentially biased in the same direction. For instance, mbMCW tests can be used to analyze bodyweights or transcript abundances determined at two different timepoints for two sets of mice that have been exposed to different conditions.

### mbMCW testing entry dataset formatting

When executing the *mbMCWtest* function, users must provide the path to a local CSV file named *X_mbMCWtest_data.csv*, where *X* serves as a user-defined identifier. *X_mbMCWtest_data.csv* can be structured in two distinct formats:

-   **Vertical layout:** This format allows appending datasets with varying structures, such as different numbers of matched-pairs per set or between each appended test. Vertical entry datasets should include the following columns:

    -   Columns *matched_condition_a* and *matched_condition_b* uniquely identify the two conditions under which matched-paired measure were collected.
    -   Column *unmatched_condition* uniquely identifies the two sets of matched-paired measures under analysis.
    -   Columns *value_a* and *value_b* contain the actual measures under analysis.
    -   As many informative columns as needed by users to contextualize the results of each test. The names of these columns should not contain the terms *condition* or *value*. While these columns are optional when running a single test, at least one column is required when running multiple tests simultaneously. All rows for each individual test must contain the same information in these columns.

-   **Horizontal layout:** This format allows appending datasets with similar structures, such as the same number of matched-paired measures collected for two conditions. Horizontal entry datasets should include the following columns:

    -   Columns *matched_condition_a* and *matched_condition_b* uniquely identify the two conditions under which matched-paired measure were collected.
    -   Columns *unmatched_condition_x* and *unmatched_condition_y* uniquely identify the two different sets of matched-paired measures under analysis.
    -   Columns *x.a.i*, *y.a.i*, *x.b.i* and *y.b.i*, where *i* represents integers to differentiate each specific matched-pair of measures, contain the actual measures under analysis.
    -   As many informative columns as needed by users to contextualize the results of each test. The name of these columns should not contain the term *condition* or have the same structure as the *x.a.i*, *y.a.i*, *x.b.i* and *y.b.i* columns. While these columns are optional when running a single test, at least one column is required when running multiple tests simultaneously.

### mbMCW testing process

The function *mbMCWtest* eliminates any matched-paired measures with at least one missing value (NA) before proceeding with the following steps.

-   To estimate the differential bias between the two sets of matched-paired measures in the dataset, the function *mbMCWtest* perfoms the following tasks:

    -   For each matched-paired measure, it subtracts the values for the two possible matched condition contrasts (*e.g.*, *a-b* and *b-a*).
    -   For each matched condition contrast, it ranks the absolute values of non-zero differences from lowest to highest. Measure pair differences with a value of 0 are assigned a 0 rank. If multiple measure pair differences have the same absolute value, all tied measure pair differences are assigned the lowest rank possible.
    -   It assigns each measure pair rank a sign based on the sign of its corresponding measure pair difference.
    -   For each set of matched-paired measures (*e.g.*, *x* and *y*), it sums the signed ranks for each matched condition contrast (*e.g.*, *a-b* and *b-a*).
    -   For each set of matched-paired measures (*e.g.*, *x* and *y*) and each matched condition contrast (*e.g.*, *a-b* and *b-a*), it calculates one mbMCW_BI. This value is obtained by dividing each sum of signed ranks by the maximum number this sum could have if the corresponding measure pairs had the highest possible positive ranks. Consequently, mbMCW_BI ranges between 1 when all the values for matched-pair measure differences in the set under analysis have the highest positive values, and -1 when all the values for matched-pair measure differences in the set under analysis have the lowest negative values.

-   To assess the significance of the mbMCW_BIs obtained from the user-provided dataset (observed mbMCW_BIs), the function *mbMCWtest* perfoms the following tasks:

    -   It generates a collection of expected-by-chance mbMCW_BIs. These expected values are obtained by rearranging the matched-pair measures between the two sets multiple times. The user-provided parameter *max_rearrangements* determines the two paths the function *mbMCWtest* can follow to generate the collection of expected-by-chance mbMCW_BIs:

        -   *mbMCW exact testing*: If the number of distinct matched-paired measure rearrangements that can alter their initial set distribution is less than *max_rearrangements*, the function *mbMCWtest* calculates mbMCW_BIs for all possible data rearrangements.
        -   *mbMCW approximated testing*: If the number of distinct matched-paired measure rearrangements that can alter their initial set distribution is greater than *max_rearrangements*, the function *mbMCWtest* performs N = *max_rearrangements* random measure rearrangements to calculate the collection of expected-by-chance mbMCW_BIs.

    -   It calculates *P~upper~* and *P~lower~* values as the fraction of expected-by-chance mbMCW_BIs that are higher or equal to and lower or equal to the observed mbMCW_BIs, respectively.

### mbMCW testing results

The *mbMCWtest* function reports to the console the total number of tests it will execute, and their exact and approximated counts. It also creates a CSV file named *X_mbMCWtest_results.csv*, where *X* is a user-defined identifier for the entry dataset CSV file. The *X_mbMCWtest_results.csv* file contains four rows for each mbMCWtest, with mbMCW_BIs calculated for each possible contrast between matched and unmatched measures (*e.g.*, *a-b*, *b-a*, *x-y* and *y-x*).

The *X_mbMCWtest_results.csv* file includes the following columns:

-   User-provided informative columns to contextualize the results of each test.
-   Columns *matched_condition_a* and *matched_condition_b* indicate the conditions for which matched-paired measures were provided.
-   Columns *unmatched_condition_x* and *unmatched_condition_y* indicate the two sets of matched-pairs measures.
-   Columns *N*, *N_x* and *N_y* indicate the total number of matched-paired measures, and their distribution between the two unmatched sets after removing any matched-pair with missing values (NAs).
-   Column *test_type* distinguishes between exact and approximated tests.
-   Column *BI_type* indicates mbMCW_BI.
-   Column *matched_condition_contrast* and *unmatched_condition_contrast* indicate the matched and unmatched condition contrast for each row of results.
-   Column *observed_BI* contains the value of mbMCW_BIs obtained from analyzing the user-provided dataset.
-   Column *expected_by_chance_BI_N* indicates the number of data rearrangements used to calculate the expected-by-chance mbMCW_BIs. This value corresponds to the lowest number between all possible measure rearrangements and the parameter *max_rearrangements*.
-   Columns *pupper* and *plower* represent the *P~upper~* and *P~lower~* values, respectively. They denote the fraction of expected-by-chance mbMCW_BIs with values higher or equal to and lower or equal to the observed mbMCW_BIs, respectively.

### Examples

#### mbMCW testing of multiple datasets organized vertically

This example of mbMCW testing entry dataset with a vertical layout showcases data for two ideal tests (Table 9). The *Contrast* I test includes data for two sets of matched-paired measures randomly selected from the same range (Figure 5). The *Contrast* II test includes data for two sets of matched-paired measures. For the first set, matched-pair measures were randomly selected from the same range. For the second set, matched-pair measures were randomly selected from two different ranges, with the measures corresponding to the second condition of each matched-pair being generally higher than the measures corresponding to the first condition of each matched-pair (Figure 5). The results of these mbMCW tests align with the intended structure of the entry datasets (Table 10).

##### Table 9. mbMCW testing entry dataset with vertical layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_vertical_mbMCWtest_data.csv", package = "MCWtests"))
```

##### Figure 5. Data structure of mbMCW testing entry dataset with vertical layout

```{r, echo = FALSE, fig.width = 6, fig.height = 5, fig.align = 'center'}
ggplot(melt(na.omit(data.table::fread(system.file("extdata", "example_vertical_mbMCWtest_data.csv", package = "MCWtests")))[
  , `:=`(replicate_pair, seq_len(.N)), by = contrast],
  id.vars = c("contrast", "replicate_pair", "unmatched_condition"),
  measure.vars = patterns("^matched_condition", "value"),
  variable.name = "pair_element",
  value.name = c("matched_condition", "value")), 
  aes(x = matched_condition, y = value, fill = unmatched_condition)) +
  geom_dotplot(binaxis = "y", stackdir = "center", aes(colour = unmatched_condition)) +
  geom_line(linewidth = 0.2, aes(group = replicate_pair, colour = unmatched_condition)) +
  facet_grid(. ~ contrast, scales = "free") +
  theme (legend.title = element_text(size = 10),
         legend.text = element_text(size = 10),
         legend.key.size = unit(0.2, "cm"),
         axis.text.x = element_text(size = 10))
```

##### Table 10. Results of mbMCW testing using an entry dataset with vertical layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_vertical_mbMCWtest_results.csv", package = "MCWtests"))
```

#### mbMCW testing of multiple datasets organized horizontally

This example of mbMCW testing entry dataset with a vertical layout showcases data for six tests (Table 11). It includes two types of informative columns to accommodate cases where users might want to perform multiple tests with a nested structure. The columns with the *contrast* prefix provide contextual information for groups of tests or rows, while the columns with the *element* prefix provide contextual information for each test or row. Each *Contrast* I row includes two sets of matched-paired measures randomly selected from the same range (Figure 6). Each *Contrast* II row includes two sets of matched-paired measures. For the first set, matched-pair measures were randomly selected from the same range. For the second set, matched-pair measures were randomly selected from two different ranges, with the measures corresponding to the second condition of each matched-pair being generally higher than the measures corresponding to the first condition of each matched-pair (Figure 6). The results of these mbMCW tests align with the intended structure of the entry datasets (Table 12).

##### Table 11. mbMCW testing entry dataset with horizontal layout

```{r, echo = FALSE, }
data.table::fread(system.file("extdata", "example_horizontal_mbMCWtest_data.csv", package = "MCWtests"))
```

##### Figure 6. Data structure of mbMCW testing entry dataset with horizontal layout

```{r, echo = FALSE, fig.width = 6, fig.height = 5, fig.align = 'center'}
ggplot(melt(na.omit(dcast(melt(data.table::fread(system.file("extdata", "example_horizontal_mbMCWtest_data.csv", package = "MCWtests")),
                                            measure.vars = patterns("\\."),
                                            variable.name = "condition",
                                            value.name = "value",
                                            value.factor = FALSE)[
                                              value != "NA", ][
                                                , `:=`(c("unmatched_condition", "matched_condition", "replicate_pair"), tstrsplit(condition, ".", fixed = TRUE))][
                                                  , `:=`(matched_condition = paste0("value_", matched_condition),
                                                         condition = NULL)],
                                       ... ~ matched_condition, value.var = "value")),
  id.vars = c("contrast", "element_ID", "replicate_pair", "unmatched_condition"),
  measure.vars = patterns("^matched_condition", "value"),
  variable.name = "pair_element",
  value.name = c("matched_condition", "value"))[
    , `:=`(unmatched_condition = fifelse(unmatched_condition == "x", "XXXX", "YYYY"))], 
  aes(x = matched_condition, y = value, fill = unmatched_condition, color = unmatched_condition)) +
  geom_dotplot(binaxis = "y", stackdir = "center", aes(colour = unmatched_condition)) +
  geom_line(linewidth = 0.2, aes(group = replicate_pair, colour = unmatched_condition)) +
  facet_grid(contrast ~ element_ID, scales = "free") +
  theme (legend.title = element_text(size = 10),
         legend.text = element_text(size = 10),
         legend.key.size = unit(0.2, "cm"),
         axis.text.x = element_text(size = 10))
```

##### Table 12. Results of mbMCW testing using an entry dataset with horizontal layout

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_horizontal_mbMCWtest_results.csv", package = "MCWtests"))
```

## 4.4 bias-measures MCW (bMCW) test

### Introduction

The bMCW test is actually a combination of two tests that assess whether a set of measures of bias for a quantitative trait between two conditions or a subset of these bias measures are themselves significantly biased in the same direction. For instance, bMCW tests can be used to analyze bias indexes obtained using other MCW tests or fold change for transcript abundances spanning the entire transcriptome or only for genes located in specific genomic regions from two sets of mice exposed to different conditions.

### bMCW testing entry dataset formatting

When executing the *bMCWtest* function, users must provide the path to a local CSV file named *X_bMCWtest_data.csv*, where *X* serves as a user-defined identifier. *X_bMCWtest_data.csv* should include the following columns:

-   Column *bias_value* contains the value of the bias measure under analysis.
-   Columns *subset_x*, where *x* represents the specific type of subset for each column, such as "chr" for chromosomes or "GO" for Gene Ontology. These columns are required if users intend to assess whether bias measures for certain subsets of elements in the dataset are significantly biased in the same direction. Columns *subset_x* can indicate whether an element belongs to a subset using either "YES" and "NO", or specific subset names like "chr1" or "chrX", or a combination of both, such as "chr1", "chrX" and "NO". The function *bMCW test* will transform the dataset to conduct independent analysis of each subset of elements marked as "YES" or with a specific subset name in each *subset_x* column.
-   As many informative columns as needed by users to contextualize the results of each test. The names of these columns should not contain the terms *bias_value* or *subset*. While these columns are optional when running a single test, at least one column is required when running multiple tests simultaneously. All rows for each individual test must contain the same information in these columns.
-   Users can specify columns with information relevant about each element or row using the column name structure *element_x*, where *x* indicates the specific information in each column (see example). However, *element_x* columns are not essential for bMCW testing and will not be included in the results file.

### bMCW testing process

The function *bMCWtest* eliminates missing values (NAs) from the dataset before proceeding with the following steps.

-   To estimate the bias for all bias measures in the entire dataset or a subset of them, the function *bMCWtest* performs the following tasks:

    -   It ranks all bias measures with non-zero values from lowest to highest. Bias measures with a value of 0 are assigned a 0 rank. If multiple bias measures have the same absolute value, all tied bias measures are asssigned the lowest rank possible.
    -   It assigns each rank a sign based on the sign of its corresponding bias measure.
    -   It calculates a whole-set bias index (bMCW_wBI) by summing the signed ranks for all elements in the dataset and dividing it by the maximum number that sum could have if all bias measures were positive. Consequently, bMCW_wBI ranges between 1 when all bias measures are positive, and -1 when all bias measures are negative.
    -   It calculates a subset bias index (bMCW_sBI) for each subset of elements under analysis by summing the signed ranks for the elements in the subset and dividing it by the maximum number that sum could have if the elements in the subset had the highest possible positive bias measures. Consequently, bMCW_sBI ranges between 1 when the bias measures for the subset in question have the highest positive bias measures in the entire dataset, and -1 when the bias measures for the subset in question have the lowest negative bias measures in the entire dataset.

-   To assess the significance of the bMCW-wBIs and bMCW-sBIs obtained from the user-provided dataset (observed bMCW-wBIs and bMCW-sBIs), the function *bMCWtest* performs the following tasks,

```         
- It generates a collection of expected-by-chance bMCW_wBIs by rearranging the signs of all signed ranks multiple times. The function *bMCWtest* also generates a collection of expected-by-chance bMCW_sBIs by rearranging the subset of elements multiple times. The user-provided parameter *max_rearrangements* determines the two paths that the function *bMCWtest* can follow to generate the collection of expected-by-chance bMCW_wBIs and bMCW_sBIs:

    - *bMCW exact testing*: If the number of distinct bias measure rearrangements that can alter their initial sign distribution or subset distribution is less than *max_rearrangements*, the function *bMCWtest* calculates bMCW_wBIs or bMCW_sBIs for all possible data rearrangements.
    - *bMCW approximated testing*: If the number of distinct bias measure rearrangements that can alter their initial sign distribution or subset distribution is greater than *max_rearrangements*, the function *bMCWtest* performs N = *max_rearrangements* random measure rearrangements to calculate the collection of expected-by-chance bMCW_wBIs or bMCW_sBIs.

- It calculates *P~upper~* and *P~lower~* values, as the fraction of expected-by-chance bMCW-wBIs and bMCW-sBIs that are higher or equal to and lower or equal to the observed bMCW-wBIs and bMCW-sBIs, respectively.
```

### bMCW testing results

The *bMCWtest* function reports to the console the total number of tests it will execute, and their exact and approximated counts. It also creates a CSV file named *X_bMCWtest_results.csv*, where *X* is a user-defined identifier for the entry dataset CSV file. The *X_bMCWtest_results.csv* file contains one row for each bMCWtest to indicate the results of whole-set bMCW testing, and as many rows as necessary to indicate the results of subset bMCW testing. Rows for whole-set analyses will be at the top of *X_bMCWtest_results.csv* file.

The *X_bMCWtest_results.csv* file includes the following columns:

-   User-provided informative columns to contextualize the results of each test.
-   Column *subset_type* indicates whether the results in each row corresponds to *whole-set* tests or specific *subset* tests, such as "chr" for chromosomes or "GO" for Gene Ontology terms.
-   Column *tested_subset* indicates the name of the subset under analysis. For *whole-set* tests, the *tested_subset* column indicates "none". For subset tests, the *tested_subset* column indicates "YES" or the specific name of the subset under analysis, such as "chr1" or "chrX".
-   Columns *N* and *n* indicate the total number of elements in the whole set and those associated with the subset under analysis, respectively, after removing missing values (NAs). For *whole-set* tests, columns *N* and *n* have the same value.
-   Column *test_type* distinguishes between exact and approximated tests.
-   Column *BI_type* indicates whether results correspond to *whole-set* tests (bMCW_wBI) or to *subset* tests (bMCW_sBIs).
-   Column *observed_BI* contains the value of bMCW_BIs obtained from analyzing the user-provided dataset.
-   Column *expected_by_chance_BI_N* indicates the number of data rearrangements used to calculate the expected-by-chance bMCW_wBIs and bMCW_sBIs. This value corresponds to the lowest number between all possible measure rearrangements and the parameter *max_rearrangements*.
-   Columns *pupper* and *plower* represent the *P~upper~* and *P~lower~* values, respectively. They denote the fraction of expected-by-chance bMCW_wBIs or bMCW_sBIs with values higher or equal to and lower or equal to the observed bMCW_wBIs or bMCW_sBIs, respectively.

### Examples

#### bMCW testing of multiple datasets

This example of bMCW testing entry dataset includes bias measures for three tests, each with ten elements distributed across four subsets: chr1, chr2, chrX and chrY (Table 13), In this example, the bias measures under study are uMCW_BIs, which range from 1 and -1. The *Contrast* I test includes bias measures randomly selected from the same range (Figure 7). The *Contrast* II test includes bias measures randomly selected from the same range, with most of them having positive values (Figure 7). The *Contrast* III includes bias measures randomly selected from the same range for three of the four subsets, while bias measures for the fourth subset are randomly selected with the largest positive values in the dataset (Figure 7). The results of these bMCW tests align with the intended structure of the entry datasets (Table 14).

##### Table 13. bMCW testing entry dataset

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_bMCWtest_data.csv", package = "MCWtests"))
```

##### Figure 7. Data structure of bMCW testing entry dataset

```{r, echo = FALSE, fig.width = 6, fig.height = 5, fig.align = 'center'}
ggplot(na.omit(data.table::fread(system.file("extdata", "example_bMCWtest_data.csv", package = "MCWtests"))), 
  aes(x = factor(element_ID, levels=unique(element_ID)), y = bias_value, fill = subset_chr)) +
  geom_dotplot(binaxis = "y", stackdir = "center", dotsize = 0.7, aes(colour = subset_chr)) +
  facet_grid(. ~ contrast, scales = "free") +
  theme (legend.title = element_text(size = 10),
         legend.text = element_text(size = 10),
         legend.key.size = unit(0.2, "cm"),
         axis.text.x = element_text(size = 10, angle = 90, hjust = 1, vjust = 0.5)) +
  labs(x = "element_ID")
```

##### Table 14. bMCW testing results

```{r, echo = FALSE}
data.table::fread(system.file("extdata", "example_bMCWtest_results.csv", package = "MCWtests"))
```

# References
