<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>MCWtests • MCWtests</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="MCWtests">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MCWtests</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/MCWtests.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/diazcastillo/MCWtests/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="MCWtests_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="MCWtests_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><script src="MCWtests_files/viz-1.8.2/viz.js"></script><link href="MCWtests_files/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="MCWtests_files/grViz-binding-1.0.11/grViz.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>MCWtests</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/diazcastillo/MCWtests/blob/main/vignettes/MCWtests.Rmd" class="external-link"><code>vignettes/MCWtests.Rmd</code></a></small>
      <div class="d-none name"><code>MCWtests.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction-to-mcw-testing">1. Introduction to MCW testing<a class="anchor" aria-label="anchor" href="#introduction-to-mcw-testing"></a>
</h2>
<p>Originally, Monte Carlo-Wilcoxon (MCW) tests were designed to
determined whether the differences between two sets of data were
significantly biased in the same direction when compared with what it
would be expected by chance. MCW tests proceed by calculating
sum-of-ranks-based bias indexes, hence the reference to Frank Wilcoxon
who invented the non-parametric rank-sum and signed-rank tests, before
and after rearranging the dataset multiple times, hence the Monte Carlo
reference often associated to analytical strategies based on repeated
random sampling<span class="citation"><sup>1–5</sup></span>.</p>
<p>The <em>MCWtests</em> package encompasses the original MCW test and
three variations that differ in the data structures and the specific
questions they interrogate.</p>
<ul>
<li><p>The <strong>matched-measures univariate MCW (muMCW)</strong>
<strong>test</strong>, the original MCW test<span class="citation"><sup>1–5</sup></span>, assesses whether one set of
inherently matched-paired measures is significantly biased in the same
direction. For instance, muMCW tests can be used to analyze bodyweights
or transcript abundances determined at two different timepoints for the
same set of mice.</p></li>
<li><p>The <strong>unmatched-measures MCW (uMCW) test</strong> assesses
whether two sets of unmatched measures and their heterogeneity are
significantly biased in the same direction. For instance, uMCW tests can
be used to analyze bodyweights or transcript abundances determined for
two sets of mice that have been maintained in different
conditions.</p></li>
<li><p>The <strong>matched-measures bivariate MCW (mbMCW) test</strong>
assesses whether two sets of inherently matched-paired measures are
significantly differentially biased in the same direction. For instance,
mbMCW tests can be used to analyze bodyweights or transcript abundances
determined at two different timepoints for two sets of mice that have
been exposed to different conditions.</p></li>
<li><p>The <strong>bias-measures MCW (bMCW) test</strong> assesses
whether a set of measures of bias for a quantitative trait between two
conditions or a subset of these bias measures are themselves
significantly biased in the same direction. For instance, bMCW tests can
be used to analyze bias indexes obtained using other MCW tests or fold
change for transcript abundances spanning the entire transcriptome or
only for genes located in specific genomic regions from two sets of mice
exposed to different conditions.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="mcw-testing-process">2. MCW testing process<a class="anchor" aria-label="anchor" href="#mcw-testing-process"></a>
</h2>
<p>Although each MCW test examines distinct data structures to address
slightly different questions, all MCW tests share two fundamental
steps:</p>
<ol style="list-style-type: decimal">
<li><p><strong>To quantitatively determine the extent and direction of
the bias of the measure under analysis</strong>, MCW tests calculate a
bias index (BI) by summing ranks and dividing these sums by the maximum
possible value of the sums. Consequently, BIs range from 1 to -1 when
the measure under analysis is completely biased in each possible
direction.</p></li>
<li><p><strong>To determine the significance of the BIs calculated for
the user-provided dataset (observed BIs)</strong>, a collection of
expected-by-chance BIs is generated by rearranging the original dataset
multiple times and calculating BIs for each iteration.
<em>P<sub>upper</sub></em> and <em>P<sub>lower</sub></em> values are
calculated as the fractions of expected-by-chance BIs that have values
higher or equal to and lower or equal to the observed BIs,
respectively.</p></li>
</ol>
<p>Each MCW test employs the user-provided parameter
<em>max_rearrangements</em> to follow two alternative paths.</p>
<ul>
<li><p><strong>MCW exact tests.</strong> If the number of distinct
rearrangements that can be generated from the dataset under analysis is
less than <em>max_rearrangements</em>, MCW tests will actually generate
all possible data rearrangements to create the collection of
expected-by-chance BIs. In this case, <em>P<sub>upper</sub></em> and
<em>P<sub>lower</sub></em> values will be exact estimations of the
likelihood of obtaining BIs with equal or more extreme values compared
to observed BIs with datasets of the same size and range but different
internal structures.</p></li>
<li><p><strong>MCW approximated tests.</strong> If the number of
distinct rearrangements that can be generated from the dataset under
analysis is greater than <em>max_rearrangements</em>, MCW tests will
perform a specified number of random data rearrangements, equal to the
value of <em>max_rearrangements</em>, to generate the collection of
expected-by-chance BIs. In this case, <em>P<sub>upper</sub></em> and
<em>P<sub>lower</sub></em> values will represent approximate estimations
of the likelihood of obtaining BIs with equal or more extreme values
compared to observed BIs with datasets of the same size and range but
different internal structures.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="mcwtests-data-and-function-structure">3. MCWtests data and function structure<a class="anchor" aria-label="anchor" href="#mcwtests-data-and-function-structure"></a>
</h2>
<p>Each MCW test involves two local files, one exported function that
users interacts with, and several internal functions that perform
specific tasks.</p>
<div class="grViz html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"diagram":"digraph flowchart {\n                  \n                  entry_dataset1 [shape = folder, label = \"entry csv dataset\", color = lightblue, style = filled]\n                  results [shape = folder, label = \"results csv table\", color = yellowgreen, style = filled]\n                  function1 [shape = box, label = \"MCWtest\", color = salmon, style = filled]\n                  function2 [shape = box, label = \"MCW_exact_test\", color = grey, style = filled]\n                  function3 [shape = box, label = \"MCW_approximated_test\", color = grey, style = filled]\n                  function4 [shape = box, label = \"MCW_exact_test_combinations\", color = grey, style = filled]\n                  function5 [shape = box, label = \"MCW_approximated_test_observed_BIs\", color = grey, style = filled]\n                  function6 [shape = box, label = \"MCW_approximated_test_simulated_BIs\", color = grey, style = filled]\n                  \n                  entry_dataset1 -> function1\n                  function1 -> results\n                  function1 -> function2\n                  function1 -> function2 [dir=\"back\"]\n                  function1 -> function3\n                  function1 -> function3 [dir=\"back\"]\n                  function2 -> function4\n                  function2 -> function4 [dir=\"back\"]\n                  function3 -> function5\n                  function3 -> function5 [dir=\"back\"]\n                  function3 -> function6\n                  function3 -> function6 [dir=\"back\"]\n                  function5 -> function6\n                  function5 -> function6 [dir=\"back\"]\n                \n                  {rank = same; entry_dataset1 results}\n                  {rank = same; function4 function5 function6}\n                  \n                  }","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><ol style="list-style-type: decimal">
<li><strong>Local files:</strong></li>
</ol>
<ul>
<li><p><strong>Entry CSV dataset</strong>. Users provide data for MCW
testing in a CSV file in the directory of their choice. This file can
contain data for a single MCW test or multiple MCW tests to be executed
simultaneously. Users can organize the entry dataset in two different
layouts, vertical or horizontal, which is particularly useful for
simultaneous testing when the data structures for each individual test
are significantly different or very similar, respectively.</p></li>
<li><p><strong>Results CSV table</strong>. MCW test functions create a
CSV file with the results of MCW testing, with a similar name to the
entry CSV dataset file, in the same directory where the entry CSV is
located.</p></li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li><strong>Exported functions:</strong></li>
</ol>
<ul>
<li>
<p><strong><em>MCWtest</em></strong> <strong>function</strong>: When
executing this function, users must specify the path to the CSV file
containing the entry dataset and the parameter
<em>max_rearrangements</em>. The <em>MCWtest</em> function interprets
this parameter as the maximum number of data rearrangement iterations
used to define the size of the collection of expected-by-chance BIs. The
<em>MCWtest</em> function performs the following tasks:</p>
<ul>
<li>It loads the entry CSV dataset.</li>
<li>It determines whether MCW tests will proceed using the exact or
approximated testing paths.</li>
<li>It calls the corresponding functions, <em>MCW_exact_test</em> and
<em>MCW_approximated_test</em>, to run the exact and approximated tests,
resepctively.</li>
<li>It writes the results of MCW testing to a CSV file in the same
directory as the entry CSV file.</li>
</ul>
</li>
</ul>
<ol start="3" style="list-style-type: decimal">
<li><strong>Internal functions:</strong></li>
</ol>
<ul>
<li>
<p><strong><em>MCW_exact_test</em></strong>
<strong>function</strong>: This function is called when the number of
distinctive data rearrangements is less than the user-provided parameter
<em>max_rearrangements</em>. The <em>MCW_exact_test</em> function
performs the following tasks:</p>
<ul>
<li>It generates all distinctive data rearrangements, either on its own
or by executing the <em>MCW_exact_test_combinations</em> function.</li>
<li>It calculates BIs for the user-provided data and each of the
distinctive rearrangements of the dataset.</li>
<li>It calculates the <em>P<sub>upper</sub></em> and
<em>P<sub>lower</sub></em> values as the fractions of expected-by-chance
BIs with values higher or equal to and lower or equal to the observed
BIs, respectively.</li>
<li>It returns the MCW exact testing results to the <em>MCWtest</em>
function.</li>
</ul>
</li>
<li><p><strong><em>MCW_exact_test_combinations</em></strong>
<strong>function</strong>: This function is called to assist the
<em>MCW_exact_test</em> function in generating all distinctive data
rearrangements.</p></li>
<li>
<p><strong><em>MCW_approximated_test</em></strong>
<strong>function</strong>: This function is called when the number of
distinctive data rearrangements is greater than the user-provided
parameter <em>max_rearrangements</em>. The
<em>MCW_approximated_test</em> function performs the following
tasks:</p>
<ul>
<li>It calls the <em>MCW_approximated_test_observed_BIs</em> function to
calculate BIs for the user-provided dataset.</li>
<li>It calls the <em>MCW_approximated_test_simulated_BIs</em> function
to randomly rearrange the data and calculate the expected-by-chance BIs
a specified number of times, which is equal to
<em>max_rearrangements</em>.</li>
<li>It calculates the <em>P<sub>upper</sub></em> and
<em>P<sub>lower</sub></em> values as the fractions of expected-by-chance
BIs with values higher or equal to and lower or equal to the observed
BIs, respectively.</li>
<li>It returns the MCW approximated testing results to the
<em>MCWtest</em> function</li>
</ul>
</li>
<li><p><strong><em>MCW_approximated_test_observed_BIs</em></strong>
<strong>function</strong>: This function is called to calculate BIs for
the user-provided dataset.</p></li>
<li><p><strong><em>MCW_approximated_test_simulated_BIs</em></strong>
<strong>function</strong>: This function is called N times, where N =
<em>max_rearrangements</em>, to randomly rearrange the data and
calculate the expected-by-chance BIs using the
<em>MCW_approximated_test_observed_BIs</em> function.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="specific-mcw-tests">4. Specific MCW tests<a class="anchor" aria-label="anchor" href="#specific-mcw-tests"></a>
</h2>
<div class="section level3">
<h3 id="unmatched-measures-mcw-umcw-test">4.1 unmatched-measures MCW (uMCW) test<a class="anchor" aria-label="anchor" href="#unmatched-measures-mcw-umcw-test"></a>
</h3>
<div class="section level4">
<h4 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h4>
<p>The uMCW test is actually a combination of two different tests that
assess whether two sets of unmatched measures and their heterogeneity
are significantly biased in the same direction. Significantly different
data heterogeneities between two conditions could indicate that the
measure under analysis is more constrained or more relaxed in one of the
conditions, potentially providing insights into the mechanisms
underlying the variation of such measure. For instance, uMCW tests can
be used to analyze bodyweights or transcript abundances determined for
two sets of mice that have been maintained in different conditions.</p>
</div>
<div class="section level4">
<h4 id="umcw-testing-entry-dataset-formatting">uMCW testing entry dataset formatting<a class="anchor" aria-label="anchor" href="#umcw-testing-entry-dataset-formatting"></a>
</h4>
<p>When executing the <em>uMCWtest</em> function, users must provide the
path to a local CSV file named <em>X_uMCWtest_data.csv</em>, where
<em>X</em> serves as a user-defined identifier.
<em>X_uMCWtest_data.csv</em> can be structured in two distinct
formats:</p>
<ul>
<li>
<p><strong>Vertical layout</strong>: This format allows appending
datasets with varying structures, such as different numbers of measures
per set or between each appended test. Vertical entry datasets should
include the following columns:</p>
<ul>
<li>The <em>condition</em> column uniquely identifies each of the two
measure sets under analysis.</li>
<li>The <em>value</em> column contains the actual measures under
analysis.</li>
<li>As many informative columns as needed by users to contextualize the
results of each test. The names of these columns should not include the
terms <em>condition</em> or <em>value</em>. While these columns are
optional when running a single test, at least one column is required
when running multiple tests simultaneously. All rows for each individual
test must contain the same information in these columns.</li>
</ul>
</li>
<li>
<p><strong>Horizontal layout</strong>: This format allows appending
datasets with similar structures, such as the same number of measures
collected for each of the two conditions. Horizontal entry datasets
should include the following columns:</p>
<ul>
<li>Columns <em>condition_a</em> and <em>condition_b</em> uniquely
identify the two measure sets under analysis.</li>
<li>Columns <em>a.i</em> and <em>b.j</em>, where <em>i</em> and
<em>j</em> represent integers to differentiate specific measures within
each set, contain the actual measures under analysis.</li>
<li>As many informative columns as needed by users to contextualize the
results of each test. The names of these columns should not contain the
term <em>condition</em> or have the same structure as the <em>a.i</em>
and <em>b.j</em> columns. While these columns are optional when running
a single test, at least one column is required when running multiple
tests simultaneously.</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="umcw-testing-process">uMCW testing process<a class="anchor" aria-label="anchor" href="#umcw-testing-process"></a>
</h4>
<p>The function <em>uMCWtest</em> eliminate missing values (NAs) from
the dataset before proceeding these steps.</p>
<ul>
<li>
<p>To estimate the bias between the two sets of measures
(<em>e.g.</em>, <em>a</em> and <em>b</em>), the function
<em>uMCWtest</em> performs these tasks:</p>
<ul>
<li>It generates all possible disjoint data pairs using measures from
both sets.</li>
<li>For each measure pair, it subtracts the second measure in the pair
from the first measure in the pair.</li>
<li>It ranks the absolute values of all non-zero measure pair
differences from lowest to highest. Measure pair differences with a
value of 0 are assigned a 0 rank. If multiple measure pair differences
have the same absolute value, all tied measure pair differences are
assigned the lowest rank possible.</li>
<li>It assigns each measure pair rank a sign based on the sign of its
corresponding measure pair difference.</li>
<li>It sums the signed ranks for measure pairs formed with measures from
the two different sets (<em>e.g.</em>, <em>a-b</em> and
<em>b-a</em>).</li>
<li>For each type of disjoint set measure pairs (<em>e.g.</em>,
<em>a-b</em> and <em>b-a</em>), it calculates uMCW_BI by dividing the
sum of signed ranks by the maximum number this sum could have if the
corresponding measure pairs had the highest possible positive ranks.
Consequently, uMCW_BI ranges between 1 when all the values for measures
in the first set are higher than all the values from measures in the
second set, and -1 when all the values for measures in the first set are
lower than all the values from measures in the second set.</li>
</ul>
</li>
<li>
<p>To estimate the bias between the heterogeneity of two sets of
measures, the function <em>uMCWtest</em> performs these tasks:</p>
<ul>
<li>It generates all possible disjoint data pairs within each set,
disregarding the order of the paired measures. For instance, the measure
pair <em>a.1-a.2</em> is considered equivalent to the measure pair
<em>a.2-a.1</em>, and only the former is retained for the subsequent
calculations.</li>
<li>For each measure pair, it subtracts the second measure from the
first measure.</li>
<li>It ranks all measure pair differences with non-zero values from
lowest to highest. Measure pair differences with a value of 0 are
assigned a 0 rank. If multiple measure pair differences have the same
absolute value, <em>uMCWtest</em> assigns all tied measure pair
differences the lowest rank possible.</li>
<li>It sums ranks for measure pairs formed with measures from the same
set (<em>e.g.</em>, <em>a-a</em> and <em>b-b</em>).</li>
<li>For each type of same-set measure pairs (<em>e.g.</em>, <em>a-a</em>
and <em>b-b</em>), it divides each sum of signed ranks by the maximum
number this sum could have if the corresponding measure pairs had the
highest possible ranks.</li>
<li>It calculates two heterogeneity bias indexes (uMCW_HBIs) by
subtracting the normalized sum of signed ranks from the previous step in
two possible directions (<em>e.g.</em>, <em>a-b</em> and <em>b-a</em>).
Consequently, uMCW_HBI ranges between 1 when at least two measures in
the first set have distinct values and all measures in the second set
have the same value, and -1 when all measures in the first set have the
same value and at least two measures in the second set have distinct
values.</li>
</ul>
</li>
<li>
<p>To assess the significance of the uMCW_BIs and uMCW_HBIs obtained
with the user-provided data (observed uMCW_BIs and uMCW_HBIs), the
function <em>uMCWtest</em> performs these tasks:</p>
<ul>
<li>
<p>It generates a collection of expected-by-chance uMCW_BIs and
uMCW_HBIs. These expected values are obtained by rearranging the
measures between the two sets multiple times. The user-provided
parameter <em>max_rearrangements</em> determines the two paths that the
function <em>uMCWtest</em> can follow to generate the collection of
expected-by-chance uMCW_BIs and uMCW_HBIs:</p>
<ul>
<li>
<em>uMCW exact testing</em>: If the number of distinct measure
rearrangements that can alter their initial set distribution is less
than <em>max_rearrangements</em>, the function <em>uMCWtest</em>
calculates uMCW_BIs and uMCW_HBIs for all possible data
rearrangements.</li>
<li>
<em>uMCW approximated testing</em>: If the number of distinct
measure rearrangements that can alter their initial set distribution is
greater than <em>max_rearrangements</em>, the function <em>uMCWtest</em>
will perform N = <em>max_rearrangements</em> random measure
rearrangements to calculate the collection of expected-by-chance
uMCW_BIs and uMCW_HBIs.</li>
</ul>
</li>
<li><p>It calculates <em>P<sub>upper</sub></em> and
<em>P<sub>lower</sub></em> values as the fraction of expected-by-chance
uMCW_BIs and uMCW_HBIs that are higher or equal to and lower or equal to
the observed uMCW_BIs and uMCW_HBIs, respectively.</p></li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="umcw-testing-results">uMCW testing results<a class="anchor" aria-label="anchor" href="#umcw-testing-results"></a>
</h4>
<p>The <em>uMCWtest</em> function reports to the console the total
number of tests it will execute, and their exact and approximated
counts. It also creates a CSV file named
<em>X_uMCWtest_results.csv</em>, where <em>X</em> is a user-defined
identifier for the entry dataset CSV file. The
<em>X_uMCWtest_results.csv</em> file contains four rows for each
uMCWtest, two for uMCW_BIs calculated for each condition contrast
(<em>e.g.</em>, <em>a-b</em> and <em>b-a</em>), and two for uMCW_HBIs
calculated for each condition contrast.</p>
<p>The <em>X_uMCWtest_results.csv</em> file includes the following
columns:</p>
<ul>
<li>User-provided informative columns to contextualize the results of
each test.</li>
<li>Columns <em>condition_a</em> and <em>condition_b</em> indicate the
two measure sets under analysis.</li>
<li>Columns <em>N</em>, <em>n_a</em> and <em>n_b</em> indicate the total
number of measures and the number of measures belonging to each set
after removing missing values (NAs).</li>
<li>Column <em>test_type</em> distinguishes between exact and
approximated tests.</li>
<li>Column <em>BI_type</em> indicates the bias index type (uMCW_BI and
uMCW_HBI) for each row of results.</li>
<li>Column <em>condition_contrast</em> indicates the set contrast
(<em>e.g.</em>, <em>a-b</em> or <em>b-a</em>) for each row of
results.</li>
<li>Column <em>observed_BI</em> contains the values of uMCW_BIs and
uMCW_HBIs obtained from analyzing the user-provided dataset.</li>
<li>Column <em>expected_by_chance_BI_N</em> indicates the number of data
rearrangements used to calculate the expected-by-chance uMCW_BIs and
uMCW_HBIs. This value corresponds to the lowest number between all
possible measure rearrangements and the parameter
<em>max_rearrangements</em>.</li>
<li>Columns <em>pupper</em> and <em>plower</em> represent the
<em>P<sub>upper</sub></em> and <em>P<sub>lower</sub></em> values,
respectively. They denote the fraction of expected-by-chance uMCW_BIs or
uMCW_HBIs with values higher or equal to and lower or equal to the
observed uMCW_BIs or uMCW_HBIs, respectively.</li>
</ul>
</div>
<div class="section level4">
<h4 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h4>
<div class="section level5">
<h5 id="umcw-testing-of-multiple-datasets-organized-vertically">uMCW testing of multiple datasets organized vertically<a class="anchor" aria-label="anchor" href="#umcw-testing-of-multiple-datasets-organized-vertically"></a>
</h5>
<p>This example of uMCW testing entry dataset with a vertical layout
showcases data for three ideal tests (Table 1). The <em>Contrast</em> I
test includes data for two sets of measures randomly selected from the
same range (Figure 1). The <em>Contrast</em> II test includes data
representing two sets of measures randomly selected from two different
ranges, with measures from the second set being generally higher than
those from the first set (Figure 1). The <em>Contrast</em> III test
includes data representing two sets of measures randomly selected from
two different ranges, with measures from the second set being generally
higher and less heterogeneous than those from the first set (Figure 1).
The results of these uMCW tests align with the intended structure of the
entry datasets (Table 2).</p>
<div class="section level6">
<h6 id="table-1--umcw-testing-entry-dataset-with-vertical-layout">Table 1. uMCW testing entry dataset with vertical layout<a class="anchor" aria-label="anchor" href="#table-1--umcw-testing-entry-dataset-with-vertical-layout"></a>
</h6>
<pre><code>    contrast condition value
      &lt;char&gt;    &lt;char&gt; &lt;int&gt;
 1:        I      AAAA    63
 2:        I      AAAA    22
 3:        I      AAAA    87
 4:        I      AAAA    34
 5:        I      AAAA    27
 6:        I      BBBB    32
 7:        I      BBBB    10
 8:        I      BBBB    74
 9:        I      BBBB    31
10:        I      BBBB    76
11:       II      AAAA    51
12:       II      AAAA     5
13:       II      AAAA    25
14:       II      AAAA     4
15:       II      AAAA    38
16:       II      BBBB    68
17:       II      BBBB    98
18:       II      BBBB    41
19:       II      BBBB   100
20:       II      BBBB    67
21:      III      AAAA    46
22:      III      AAAA    14
23:      III      AAAA    35
24:      III      AAAA    11
25:      III      AAAA    36
26:      III      BBBB    56
27:      III      BBBB    51
28:      III      BBBB    50
29:      III      BBBB    51
30:      III      BBBB    60
    contrast condition value</code></pre>
</div>
<div class="section level6">
<h6 id="figure-1--data-structure-of-umcw-testing-entry-dataset-with-vertical-layout">Figure 1. Data structure of uMCW testing entry dataset with vertical
layout<a class="anchor" aria-label="anchor" href="#figure-1--data-structure-of-umcw-testing-entry-dataset-with-vertical-layout"></a>
</h6>
<p><img src="MCWtests_files/figure-html/unnamed-chunk-4-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="table-2--results-of-umcw-testing-using-an-entry-dataset-with-vertical-layout">Table 2. Results of uMCW testing using an entry dataset with
vertical layout<a class="anchor" aria-label="anchor" href="#table-2--results-of-umcw-testing-using-an-entry-dataset-with-vertical-layout"></a>
</h6>
<pre><code>    contrast condition_a condition_b     N   n_a   n_b    test_type  BI_type
      &lt;char&gt;      &lt;char&gt;      &lt;char&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;       &lt;char&gt;   &lt;char&gt;
 1:        I        AAAA        BBBB    10     5     5 approximated  uMCW_BI
 2:        I        AAAA        BBBB    10     5     5 approximated  uMCW_BI
 3:        I        AAAA        BBBB    10     5     5 approximated uMCW_HBI
 4:        I        AAAA        BBBB    10     5     5 approximated uMCW_HBI
 5:       II        AAAA        BBBB    10     5     5 approximated  uMCW_BI
 6:       II        AAAA        BBBB    10     5     5 approximated  uMCW_BI
 7:       II        AAAA        BBBB    10     5     5 approximated uMCW_HBI
 8:       II        AAAA        BBBB    10     5     5 approximated uMCW_HBI
 9:      III        AAAA        BBBB    10     5     5 approximated  uMCW_BI
10:      III        AAAA        BBBB    10     5     5 approximated  uMCW_BI
11:      III        AAAA        BBBB    10     5     5 approximated uMCW_HBI
12:      III        AAAA        BBBB    10     5     5 approximated uMCW_HBI
    condition_contrast observed_BI expected_by_chance_BI_N pupper plower
                &lt;char&gt;       &lt;num&gt;                   &lt;int&gt;  &lt;num&gt;  &lt;num&gt;
 1:          AAAA-BBBB  0.03948718                     200  0.430  0.595
 2:          BBBB-AAAA -0.03948718                     200  0.595  0.430
 3:          AAAA-BBBB -0.02857143                     200  0.500  0.505
 4:          BBBB-AAAA  0.02857143                     200  0.505  0.500
 5:          AAAA-BBBB -0.69794872                     200  1.000  0.005
 6:          BBBB-AAAA  0.69794872                     200  0.005  1.000
 7:          AAAA-BBBB -0.07766990                     200  0.720  0.285
 8:          BBBB-AAAA  0.07766990                     200  0.285  0.720
 9:          AAAA-BBBB -0.68358974                     200  1.000  0.000
10:          BBBB-AAAA  0.68358974                     200  0.000  1.000
11:          AAAA-BBBB  0.39130435                     200  0.040  0.970
12:          BBBB-AAAA -0.39130435                     200  0.970  0.040</code></pre>
</div>
</div>
<div class="section level5">
<h5 id="umcw-testing-of-multiple-datasets-organized-horizontally">uMCW testing of multiple datasets organized horizontally<a class="anchor" aria-label="anchor" href="#umcw-testing-of-multiple-datasets-organized-horizontally"></a>
</h5>
<p>This example of uMCW testing entry dataset with a horizontal layout
showcases data for nine ideal tests (Table 3). It includes two types of
informative columns to accommodate cases where users might want to
perform multiple tests with a nested structure. The columns with the
<em>contrast</em> prefix provide contextual information for groups of
tests or rows, while the columns with the <em>element</em> prefix
provide contextual information for each test or row. Each
<em>Contrast</em> I row includes two sets of measures randomly selected
from the same range (Figure 2). Each <em>Contrast</em> II row includes
two sets of measures randomly selected from two different ranges, with
the measures from the second set being generally higher than those from
the first set (Figure 2). Each <em>Contrast</em> III row includes two
sets of measures randomly selected from two different ranges, with the
measures from the second set being generally higher and less
heterogeneous than those from the first set (Figure 1). The results of
these uMCW tests align with the intended structure of the entry datasets
(Table 4).</p>
<div class="section level6">
<h6 id="table-3--umcw-testing-entry-dataset-with-horizontal-layout">Table 3. uMCW testing entry dataset with horizontal layout<a class="anchor" aria-label="anchor" href="#table-3--umcw-testing-entry-dataset-with-horizontal-layout"></a>
</h6>
<pre><code>   contrast contrast_trait element_ID element_chr element_start element_end
     &lt;char&gt;         &lt;char&gt;     &lt;char&gt;       &lt;int&gt;         &lt;int&gt;       &lt;int&gt;
1:        I        trait_a         x1           1          1000        2000
2:        I        trait_a         x2           1          5000        5500
3:        I        trait_a         x3           1         90000      100000
4:       II        trait_b         x1           1          1000        2000
5:       II        trait_b         x2           1          5000        5500
6:       II        trait_b         x3           1         90000      100000
7:      III        trait_b         x1           1          1000        2000
8:      III        trait_b         x2           1          5000        5500
9:      III        trait_b         x3           1         90000      100000
   condition_a condition_b   a.1   a.2   a.3   a.4   a.5   b.1   b.2   b.3
        &lt;char&gt;      &lt;char&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
1:        AAAA        BBBB    91    56    65    63    23    17    NA    50
2:        AAAA        BBBB    96    45    42    29    12    25    36    28
3:        AAAA        BBBB    77    98    31    93    34    34    97    77
4:        AAAA        BBBB    54    23    59    28    52    90    92    91
5:        AAAA        BBBB    37     5    18    26    11    48    84    62
6:        AAAA        BBBB    28    35    43    13     6    60    71    77
7:        AAAA        BBBB    20    32    14     3     6    46    56    45
8:        AAAA        BBBB    50    38    28    47    27    55    46    51
9:        AAAA        BBBB    31    36     2    37    47    54    48    48
     b.4   b.5
   &lt;int&gt; &lt;int&gt;
1:    39    30
2:    79    82
3:    53    11
4:    50    68
5:    86    53
6:    89    97
7:    57    50
8:    46    60
9:    46    48</code></pre>
</div>
<div class="section level6">
<h6 id="figure-2--data-structure-of-umcw-testing-entry-dataset-with-horizontal-layout">Figure 2. Data structure of uMCW testing entry dataset with
horizontal layout<a class="anchor" aria-label="anchor" href="#figure-2--data-structure-of-umcw-testing-entry-dataset-with-horizontal-layout"></a>
</h6>
<p><img src="MCWtests_files/figure-html/unnamed-chunk-7-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="table-4--results-of-umcw-testing-using-an-entry-dataset-with-horizontal-layout">Table 4. Results of uMCW testing using an entry dataset with
horizontal layout<a class="anchor" aria-label="anchor" href="#table-4--results-of-umcw-testing-using-an-entry-dataset-with-horizontal-layout"></a>
</h6>
<pre><code>    contrast contrast_trait element_ID element_chr element_start element_end
      &lt;char&gt;         &lt;char&gt;     &lt;char&gt;       &lt;int&gt;         &lt;int&gt;       &lt;int&gt;
 1:        I        trait_a         x1           1          1000        2000
 2:        I        trait_a         x1           1          1000        2000
 3:        I        trait_a         x1           1          1000        2000
 4:        I        trait_a         x1           1          1000        2000
 5:        I        trait_a         x2           1          5000        5500
 6:        I        trait_a         x2           1          5000        5500
 7:        I        trait_a         x2           1          5000        5500
 8:        I        trait_a         x2           1          5000        5500
 9:        I        trait_a         x3           1         90000      100000
10:        I        trait_a         x3           1         90000      100000
11:        I        trait_a         x3           1         90000      100000
12:        I        trait_a         x3           1         90000      100000
13:       II        trait_b         x1           1          1000        2000
14:       II        trait_b         x1           1          1000        2000
15:       II        trait_b         x1           1          1000        2000
16:       II        trait_b         x1           1          1000        2000
17:       II        trait_b         x2           1          5000        5500
18:       II        trait_b         x2           1          5000        5500
19:       II        trait_b         x2           1          5000        5500
20:       II        trait_b         x2           1          5000        5500
21:       II        trait_b         x3           1         90000      100000
22:       II        trait_b         x3           1         90000      100000
23:       II        trait_b         x3           1         90000      100000
24:       II        trait_b         x3           1         90000      100000
25:      III        trait_b         x1           1          1000        2000
26:      III        trait_b         x1           1          1000        2000
27:      III        trait_b         x1           1          1000        2000
28:      III        trait_b         x1           1          1000        2000
29:      III        trait_b         x2           1          5000        5500
30:      III        trait_b         x2           1          5000        5500
31:      III        trait_b         x2           1          5000        5500
32:      III        trait_b         x2           1          5000        5500
33:      III        trait_b         x3           1         90000      100000
34:      III        trait_b         x3           1         90000      100000
35:      III        trait_b         x3           1         90000      100000
36:      III        trait_b         x3           1         90000      100000
    contrast contrast_trait element_ID element_chr element_start element_end
    condition_a condition_b     N   n_a   n_b    test_type  BI_type
         &lt;char&gt;      &lt;char&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;       &lt;char&gt;   &lt;char&gt;
 1:        AAAA        BBBB     9     5     4        exact  uMCW_BI
 2:        AAAA        BBBB     9     5     4        exact uMCW_HBI
 3:        AAAA        BBBB     9     5     4        exact  uMCW_BI
 4:        AAAA        BBBB     9     5     4        exact uMCW_HBI
 5:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
 6:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
 7:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
 8:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
 9:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
10:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
11:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
12:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
13:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
14:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
15:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
16:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
17:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
18:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
19:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
20:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
21:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
22:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
23:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
24:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
25:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
26:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
27:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
28:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
29:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
30:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
31:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
32:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
33:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
34:        AAAA        BBBB    10     5     5 approximated  uMCW_BI
35:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
36:        AAAA        BBBB    10     5     5 approximated uMCW_HBI
    condition_a condition_b     N   n_a   n_b    test_type  BI_type
    condition_contrast  observed_BI expected_by_chance_BI_N     pupper
                &lt;char&gt;        &lt;num&gt;                   &lt;int&gt;      &lt;num&gt;
 1:          AAAA-BBBB  0.500800000                     126 0.06349206
 2:          AAAA-BBBB  0.402985075                     126 0.25396825
 3:          BBBB-AAAA -0.500800000                     126 0.94444444
 4:          BBBB-AAAA -0.402985075                     126 0.76984127
 5:          AAAA-BBBB -0.066153846                     200 0.58000000
 6:          BBBB-AAAA  0.066153846                     200 0.42500000
 7:          AAAA-BBBB  0.064039409                     200 0.41000000
 8:          BBBB-AAAA -0.064039409                     200 0.59500000
 9:          AAAA-BBBB  0.161538462                     200 0.31500000
10:          BBBB-AAAA -0.161538462                     200 0.70500000
11:          AAAA-BBBB -0.043062201                     200 0.54500000
12:          BBBB-AAAA  0.043062201                     200 0.47000000
13:          AAAA-BBBB -0.649743590                     200 0.98000000
14:          BBBB-AAAA  0.649743590                     200 0.02000000
15:          AAAA-BBBB  0.024390244                     200 0.46500000
16:          BBBB-AAAA -0.024390244                     200 0.53500000
17:          AAAA-BBBB -0.778974359                     200 1.00000000
18:          BBBB-AAAA  0.778974359                     200 0.00500000
19:          AAAA-BBBB -0.123809524                     200 0.85500000
20:          BBBB-AAAA  0.123809524                     200 0.15500000
21:          AAAA-BBBB -0.801538462                     200 1.00000000
22:          BBBB-AAAA  0.801538462                     200 0.01000000
23:          AAAA-BBBB  0.009803922                     200 0.46500000
24:          BBBB-AAAA -0.009803922                     200 0.54500000
25:          AAAA-BBBB -0.807692308                     200 1.00000000
26:          BBBB-AAAA  0.807692308                     200 0.00000000
27:          AAAA-BBBB  0.303921569                     200 0.00000000
28:          BBBB-AAAA -0.303921569                     200 1.00000000
29:          AAAA-BBBB -0.558461538                     200 0.98500000
30:          BBBB-AAAA  0.558461538                     200 0.01500000
31:          AAAA-BBBB  0.288888889                     200 0.07000000
32:          BBBB-AAAA -0.288888889                     200 0.93000000
33:          AAAA-BBBB -0.576923077                     200 0.99000000
34:          BBBB-AAAA  0.576923077                     200 0.01500000
35:          AAAA-BBBB  0.527777778                     200 0.01500000
36:          BBBB-AAAA -0.527777778                     200 0.99000000
    condition_contrast  observed_BI expected_by_chance_BI_N     pupper
        plower
         &lt;num&gt;
 1: 0.94444444
 2: 0.76984127
 3: 0.06349206
 4: 0.25396825
 5: 0.42500000
 6: 0.58000000
 7: 0.59500000
 8: 0.41000000
 9: 0.70500000
10: 0.31500000
11: 0.47000000
12: 0.54500000
13: 0.02000000
14: 0.98000000
15: 0.53500000
16: 0.46500000
17: 0.00500000
18: 1.00000000
19: 0.15500000
20: 0.85500000
21: 0.01000000
22: 1.00000000
23: 0.54500000
24: 0.46500000
25: 0.00000000
26: 1.00000000
27: 1.00000000
28: 0.00000000
29: 0.01500000
30: 0.98500000
31: 0.93000000
32: 0.07000000
33: 0.01500000
34: 0.99000000
35: 0.99000000
36: 0.01500000
        plower</code></pre>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="matched-measures-univariate-mcw-mumcw-test">4.2 matched-measures univariate MCW (muMCW) test<a class="anchor" aria-label="anchor" href="#matched-measures-univariate-mcw-mumcw-test"></a>
</h3>
<div class="section level4">
<h4 id="introduction-1">Introduction<a class="anchor" aria-label="anchor" href="#introduction-1"></a>
</h4>
<p>The muMCW test assesses whether one set of inherently matched-paired
measures is significantly biased in the same direction. For instance,
muMCW tests can be used to analyze bodyweights or transcript abundances
determined at two different timepoints for the same set of mice.</p>
</div>
<div class="section level4">
<h4 id="mumcw-testing-entry-dataset-formatting">muMCW testing entry dataset formatting<a class="anchor" aria-label="anchor" href="#mumcw-testing-entry-dataset-formatting"></a>
</h4>
<p>When executing the <em>muMCWtest</em> function, users must provide
the path to a local CSV file named <em>X_muMCWtest_data.csv</em>, where
<em>X</em> serves as a user-defined identifier.
<em>X_muMCWtest_data.csv</em> can be structured in two distinct
formats:</p>
<ul>
<li>
<p><strong>Vertical layout:</strong> This format allows appending
datasets with varying structures, such as different numbers of measure
matched-pairs for each appended test. Vertical entry datasets should
include the following columns:</p>
<ul>
<li>Columns <em>condition_a</em> and <em>condition_b</em> uniquely
identify the two conditions under which matched-paired measures were
collected.</li>
<li>Columns <em>value_a</em> and <em>value_b</em> contain the actual
measures under analysis.</li>
<li>As many informative columns as needed by users to contextualize the
results of each test. The names of the these columns should not contain
the terms <em>condition</em> or <em>value</em>. While these columns are
optional when running a single test, at least one column is required
when running multiple tests simultaneously. All rows for each individual
test must contain the same information in these columns.</li>
</ul>
</li>
<li>
<p><strong>Horizontal layout:</strong> This format allows appending
datasets with similar structures, such as the same number of
matched-paired measures for each appended test. Horizontal entry
datasets should include the following columns:</p>
<ul>
<li>Columns <em>condition_a</em> and <em>condition_b</em> uniquely
identify the two conditions under which matched-paired measures were
collected.</li>
<li>Columns <em>a.i</em> and <em>b.i</em>, where <em>i</em> represents
integers to differentiate each specific matched-pairs of measures,
contain the actual measures under analysis.</li>
<li>As many informative columns as needed by users to contextualize the
results of each test. The names of these columns should not contain the
term <em>condition</em> or have the same structure as the <em>a.i</em>
and <em>b.i</em> columns. While these columns are optional when running
a single test, at least one column is required when running multiple
tests simultaneously.</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="mumcw-testing-process">muMCW testing process<a class="anchor" aria-label="anchor" href="#mumcw-testing-process"></a>
</h4>
<p>The function <em>muMCWtest</em> eliminates any matched-paired
measures with at least one missing value (NA) before proceeding with the
following steps.</p>
<ul>
<li>
<p>To estimate the bias for all matched-paired measures in the
dataset, the function <em>muMCWtest</em> performs the following
tasks:</p>
<ul>
<li>For each matched-pair of measures, it subtracts values for the two
possible condition contrasts (<em>e.g.</em>, <em>a-b</em> and
<em>b-a</em>).</li>
<li>For each condition contrast, it ranks the absolute values of
non-zero differences from lowest to highest. Measure pair differences
with a value of 0 are assigned a 0 rank. If multiple measure pair
differences have the same absolute value, all tied measure pair
differences are assigned the lowest rank possible.</li>
<li>It assigns each measure pair rank a sign based on the sign of its
corresponding measure pair difference.</li>
<li>It sums the signed ranks for each condition contrast.</li>
<li>It calculates muMCW_BI by dividing each sum of signed ranks by the
maximum number that sum could have if the corresponding measure pairs
had the highest possible positive ranks. Consequently, muMCW_BI ranges
between 1 when all measures corresponding to the first condition are
higher than all measures corresponding to the second condition, and -1
when all measures corresponding to the first condition are lower than
all measures corresponding to the second condition.</li>
</ul>
</li>
<li>
<p>To assess the significance of the muMCW_BIs obtained from the
user-provided dataset (observed muMCW_BIs), the function
<em>muMCWtest</em> performs the following tasks:</p>
<ul>
<li>
<p>It generates a collection of expected-by-chance muMCW_BIs. These
expected values are obtained by rearranging the measures between and
within the two conditions multiple times. The user-provided parameter
<em>max_rearrangements</em> determines the two paths that the function
<em>muMCWtest</em> can follow to generate the collection of
expected-by-chance muMCW_BIs:</p>
<ul>
<li>
<em>muMCW exact testing</em>: If the number of distinct measure
rearrangements that can alter their initial pair and set distribution is
less than <em>max_rearrangements</em>, the function <em>muMCWtest</em>
calculates muMCW_BIs for all possible data rearrangements.</li>
<li>
<em>muMCW approximated testing</em>: If the number of distinct
measure rearrangements that can alter their initial pair and set
distribution is greater than <em>max_rearrangements</em>, the function
<em>muMCWtest</em> performs N = <em>max_rearrangements</em> random
measure rearrangements to calculate the collection of expected-by-chance
muMCW_BIs.</li>
</ul>
</li>
<li><p>It calculates the <em>P<sub>upper</sub></em> and
<em>P<sub>lower</sub></em> values as the fraction of expected-by-chance
muMCW_BIs that are higher or equal to and lower or equal to the observed
muMCW_BIs, respectively.</p></li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="mumcw-testing-results">muMCW testing results<a class="anchor" aria-label="anchor" href="#mumcw-testing-results"></a>
</h4>
<p>The <em>muMCWtest</em> function reports to the console the total
number of tests it will execute, and their exact and approximated
counts. It also creates a CSV file named
<em>X_muMCWtest_results.csv</em> where <em>X</em> is a user-defined
identifier for the entry dataset CSV file. The
<em>X_muMCWtest_results.csv</em> file contains two rows for each muMCW
test, with muMCW_BIs calculated for each possible condition contrast
(<em>e.g.</em>, <em>a-b</em> and <em>b-a</em>).</p>
<p>The <em>X_muMCWtest_results.csv</em> file includes the following
columns:</p>
<ul>
<li>User-provided informative columns to contextualize the results of
each test.</li>
<li>Columns <em>condition_a</em> and <em>condition_b</em> indicate the
two conditions for which matched-paired measures were provided.</li>
<li>Column <em>N</em> indicates the total number of measure
matched-pairs after removing matched-pairs with missing values
(NAs).</li>
<li>Column <em>test_type</em> distinguishes between exact and
approximated tests.</li>
<li>Column <em>BI_type</em> indicates muMCW_BI.</li>
<li>Column <em>condition_contrast</em> indicates the condition contrast
for each row of results.</li>
<li>Column <em>observed_BI</em> contains the value of muMCW_BIs obtained
from analyzing the user-provided dataset.</li>
<li>Column <em>expected_by_chance_BI_N</em> indicates the number of data
rearrangements used to calculate the expected-by-chance muMCW_BIs. This
value corresponds to the lowest number between all possible measure
rearrangements and the parameter <em>max_rearrangements</em>.</li>
<li>Columns <em>pupper</em> and <em>plower</em> represent
<em>P<sub>upper</sub></em> and <em>P<sub>lower</sub></em> values,
respectively. They denote the fraction of expected-by-chance muMCW_BIs
with values higher or equal to and lower or equal to the observed
muMCW_BIs, respectively.</li>
</ul>
</div>
<div class="section level4">
<h4 id="examples-1">Examples<a class="anchor" aria-label="anchor" href="#examples-1"></a>
</h4>
<div class="section level5">
<h5 id="mumcw-testing-of-multiple-datasets-organized-vertically">muMCW testing of multiple datasets organized vertically<a class="anchor" aria-label="anchor" href="#mumcw-testing-of-multiple-datasets-organized-vertically"></a>
</h5>
<p>This example of muMCW testing entry dataset with a vertical layout
showcases data for two ideal tests (Table 5). The <em>Contrast</em> I
test includes data representing one set of matched-paired measures
randomly selected from the same range (Figure 3). The <em>Contrast</em>
II test includes data representing one set of matched-paired measures
randomly selected from two different ranges, with measures from the
second condition in each matched-pair being generally higher than
measures from the first condition in each matched-pair (Figure 3). The
results of these muMCW tests align with the intended structure of the
entry datasets (Table 6).</p>
<div class="section level6">
<h6 id="table-5--mumcw-testing-entry-dataset-with-vertical-layout">Table 5. muMCW testing entry dataset with vertical layout<a class="anchor" aria-label="anchor" href="#table-5--mumcw-testing-entry-dataset-with-vertical-layout"></a>
</h6>
<pre><code>    contrast condition_a condition_b value_a value_b
      &lt;char&gt;      &lt;char&gt;      &lt;char&gt;   &lt;int&gt;   &lt;int&gt;
 1:        I        AAAA        BBBB      19      35
 2:        I        AAAA        BBBB      81      43
 3:        I        AAAA        BBBB      43       2
 4:        I        AAAA        BBBB      96      99
 5:        I        AAAA        BBBB      33      69
 6:        I        AAAA        BBBB      65      53
 7:        I        AAAA        BBBB      96      43
 8:        I        AAAA        BBBB      15      37
 9:        I        AAAA        BBBB      92      82
10:        I        AAAA        BBBB      46      56
11:       II        AAAA        BBBB      25      52
12:       II        AAAA        BBBB       6      83
13:       II        AAAA        BBBB      56      66
14:       II        AAAA        BBBB      33      40
15:       II        AAAA        BBBB      28      47
16:       II        AAAA        BBBB      28      48
17:       II        AAAA        BBBB      19      67
18:       II        AAAA        BBBB      27      55
19:       II        AAAA        BBBB      47      45
20:       II        AAAA        BBBB      57      77
    contrast condition_a condition_b value_a value_b</code></pre>
</div>
<div class="section level6">
<h6 id="figure-3--data-structure-of-mumcw-testing-entry-dataset-with-vertical-layout">Figure 3. Data structure of muMCW testing entry dataset with
vertical layout<a class="anchor" aria-label="anchor" href="#figure-3--data-structure-of-mumcw-testing-entry-dataset-with-vertical-layout"></a>
</h6>
<p><img src="MCWtests_files/figure-html/unnamed-chunk-10-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="table-6--results-of-mumcw-testing-using-an-entry-dataset-with-vertical-layout">Table 6. Results of muMCW testing using an entry dataset with
vertical layout<a class="anchor" aria-label="anchor" href="#table-6--results-of-mumcw-testing-using-an-entry-dataset-with-vertical-layout"></a>
</h6>
<pre><code>   contrast condition_a condition_b     N    test_type  BI_type
     &lt;char&gt;      &lt;char&gt;      &lt;char&gt; &lt;int&gt;       &lt;char&gt;   &lt;char&gt;
1:        I        AAAA        BBBB    10 approximated muMCW_BI
2:        I        AAAA        BBBB    10 approximated muMCW_BI
3:       II        AAAA        BBBB    10 approximated muMCW_BI
4:       II        AAAA        BBBB    10 approximated muMCW_BI
   condition_contrast observed_BI expected_by_chance_BI_N pupper plower
               &lt;char&gt;       &lt;num&gt;                   &lt;int&gt;  &lt;num&gt;  &lt;num&gt;
1:          AAAA-BBBB   0.2181818                     200  0.285  0.725
2:          BBBB-AAAA  -0.2181818                     200  0.725  0.285
3:          AAAA-BBBB  -0.9454545                     200  1.000  0.000
4:          BBBB-AAAA   0.9454545                     200  0.000  1.000</code></pre>
</div>
</div>
<div class="section level5">
<h5 id="mumcw-testing-of-multiple-datasets-organized-horizontally">muMCW testing of multiple datasets organized horizontally<a class="anchor" aria-label="anchor" href="#mumcw-testing-of-multiple-datasets-organized-horizontally"></a>
</h5>
<p>This example of muMCW testing entry dataset with a horizontal layout
showcases data for six ideal tests (Table 7). It includes two types of
informative columns to accommodate cases where users might want to
perform multiple tests with a nested structure. The columns with the
<em>contrast</em> prefix provide contextual information for groups of
tests or rows, while the columns with the <em>element</em> prefix
provide contextual information for each test or row. Each
<em>Contrast</em> I row includes one set of matched-paired measures
randomly selected from the same range (Figure 4). Each <em>Contrast</em>
II row includes one set of matched-paired measures randomly selected
from two different ranges, with the measures for the second condition in
each matched-pair being generally higher than measures from the first
condition in each matched-pair (Figure 4). The results of these muMCW
tests align with the intended structure of the entry datasets (Table
8).</p>
<div class="section level6">
<h6 id="table-7--mumcw-testing-entry-dataset-with-horizontal-layout">Table 7. muMCW testing entry dataset with horizontal layout<a class="anchor" aria-label="anchor" href="#table-7--mumcw-testing-entry-dataset-with-horizontal-layout"></a>
</h6>
<pre><code>   contrast contrast_trait element_ID element_chr element_start element_end
     &lt;char&gt;         &lt;char&gt;     &lt;char&gt;       &lt;int&gt;         &lt;int&gt;       &lt;int&gt;
1:        I        trait_a         x1           1          1000        2000
2:        I        trait_a         x2           1          5000        5500
3:        I        trait_a         x3           1         90000      100000
4:       II        trait_b         x1           1          1000        2000
5:       II        trait_b         x2           1          5000        5500
6:       II        trait_b         x3           1         90000      100000
   condition_a condition_b   a.1   a.2   a.3   a.4   a.5   b.1   b.2   b.3
        &lt;char&gt;      &lt;char&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
1:        AAAA        BBBB    91    56    65    63    23    17    57    NA
2:        AAAA        BBBB    96    45    42    29    12    25    36    28
3:        AAAA        BBBB    77    98    31    93    34    34    97    77
4:        AAAA        BBBB    54    23    59    28    52    90    92    91
5:        AAAA        BBBB    37     5    18    26    11    48    84    62
6:        AAAA        BBBB    28    35    43    13     6    60    71    77
     b.4   b.5
   &lt;int&gt; &lt;int&gt;
1:    39    30
2:    79    82
3:    53    11
4:    50    68
5:    86    53
6:    89    97</code></pre>
</div>
<div class="section level6">
<h6 id="figure-4--data-structure-of-mumcw-testing-entry-dataset-with-horizontal-layout">Figure 4. Data structure of muMCW testing entry dataset with
horizontal layout<a class="anchor" aria-label="anchor" href="#figure-4--data-structure-of-mumcw-testing-entry-dataset-with-horizontal-layout"></a>
</h6>
<p><img src="MCWtests_files/figure-html/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="table-8--results-of-mumcw-testing-using-an-entry-dataset-with-horizontal-layout">Table 8. Results of muMCW testing using an entry dataset with
horizontal layout<a class="anchor" aria-label="anchor" href="#table-8--results-of-mumcw-testing-using-an-entry-dataset-with-horizontal-layout"></a>
</h6>
<pre><code>    contrast contrast_trait element_ID element_chr element_start element_end
      &lt;char&gt;         &lt;char&gt;     &lt;char&gt;       &lt;int&gt;         &lt;int&gt;       &lt;int&gt;
 1:        I        trait_a         x1           1          1000        2000
 2:        I        trait_a         x1           1          1000        2000
 3:        I        trait_a         x2           1          5000        5500
 4:        I        trait_a         x2           1          5000        5500
 5:        I        trait_a         x3           1         90000      100000
 6:        I        trait_a         x3           1         90000      100000
 7:       II        trait_b         x1           1          1000        2000
 8:       II        trait_b         x1           1          1000        2000
 9:       II        trait_b         x2           1          5000        5500
10:       II        trait_b         x2           1          5000        5500
11:       II        trait_b         x3           1         90000      100000
12:       II        trait_b         x3           1         90000      100000
    condition_a condition_b     N    test_type  BI_type condition_contrast
         &lt;char&gt;      &lt;char&gt; &lt;int&gt;       &lt;char&gt;   &lt;char&gt;             &lt;char&gt;
 1:        AAAA        BBBB     4        exact muMCW_BI          AAAA-BBBB
 2:        AAAA        BBBB     4        exact muMCW_BI          BBBB-AAAA
 3:        AAAA        BBBB     5 approximated muMCW_BI          AAAA-BBBB
 4:        AAAA        BBBB     5 approximated muMCW_BI          BBBB-AAAA
 5:        AAAA        BBBB     5 approximated muMCW_BI          AAAA-BBBB
 6:        AAAA        BBBB     5 approximated muMCW_BI          BBBB-AAAA
 7:        AAAA        BBBB     5 approximated muMCW_BI          AAAA-BBBB
 8:        AAAA        BBBB     5 approximated muMCW_BI          BBBB-AAAA
 9:        AAAA        BBBB     5 approximated muMCW_BI          AAAA-BBBB
10:        AAAA        BBBB     5 approximated muMCW_BI          BBBB-AAAA
11:        AAAA        BBBB     5 approximated muMCW_BI          AAAA-BBBB
12:        AAAA        BBBB     5 approximated muMCW_BI          BBBB-AAAA
    observed_BI expected_by_chance_BI_N    pupper    plower
          &lt;num&gt;                   &lt;int&gt;     &lt;num&gt;     &lt;num&gt;
 1:  0.40000000                     105 0.7142857 0.6380952
 2: -0.40000000                     105 0.6380952 0.7142857
 3:  0.06666667                     200 0.4650000 0.6300000
 4: -0.06666667                     200 0.6300000 0.4650000
 5:  0.33333333                     200 0.3050000 0.7700000
 6: -0.33333333                     200 0.7700000 0.3050000
 7: -1.00000000                     200 1.0000000 0.0500000
 8:  1.00000000                     200 0.0500000 1.0000000
 9: -1.00000000                     200 1.0000000 0.0650000
10:  1.00000000                     200 0.0650000 1.0000000
11: -1.00000000                     200 1.0000000 0.0250000
12:  1.00000000                     200 0.0250000 1.0000000</code></pre>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="matched-measures-bivariate-mcw-mbmcw-test">4.3 matched-measures bivariate MCW (mbMCW) test<a class="anchor" aria-label="anchor" href="#matched-measures-bivariate-mcw-mbmcw-test"></a>
</h3>
<div class="section level4">
<h4 id="introduction-2">Introduction<a class="anchor" aria-label="anchor" href="#introduction-2"></a>
</h4>
<p>The mbMCW test assesses whether two sets of inherently matched-paired
measures are significantly differentially biased in the same direction.
For instance, mbMCW tests can be used to analyze bodyweights or
transcript abundances determined at two different timepoints for two
sets of mice that have been exposed to different conditions.</p>
</div>
<div class="section level4">
<h4 id="mbmcw-testing-entry-dataset-formatting">mbMCW testing entry dataset formatting<a class="anchor" aria-label="anchor" href="#mbmcw-testing-entry-dataset-formatting"></a>
</h4>
<p>When executing the <em>mbMCWtest</em> function, users must provide
the path to a local CSV file named <em>X_mbMCWtest_data.csv</em>, where
<em>X</em> serves as a user-defined identifier.
<em>X_mbMCWtest_data.csv</em> can be structured in two distinct
formats:</p>
<ul>
<li>
<p><strong>Vertical layout:</strong> This format allows appending
datasets with varying structures, such as different numbers of
matched-pairs per set or between each appended test. Vertical entry
datasets should include the following columns:</p>
<ul>
<li>Columns <em>matched_condition_a</em> and
<em>matched_condition_b</em> uniquely identify the two conditions under
which matched-paired measure were collected.</li>
<li>Column <em>unmatched_condition</em> uniquely identifies the two sets
of matched-paired measures under analysis.</li>
<li>Columns <em>value_a</em> and <em>value_b</em> contain the actual
measures under analysis.</li>
<li>As many informative columns as needed by users to contextualize the
results of each test. The names of these columns should not contain the
terms <em>condition</em> or <em>value</em>. While these columns are
optional when running a single test, at least one column is required
when running multiple tests simultaneously. All rows for each individual
test must contain the same information in these columns.</li>
</ul>
</li>
<li>
<p><strong>Horizontal layout:</strong> This format allows appending
datasets with similar structures, such as the same number of
matched-paired measures collected for two conditions. Horizontal entry
datasets should include the following columns:</p>
<ul>
<li>Columns <em>matched_condition_a</em> and
<em>matched_condition_b</em> uniquely identify the two conditions under
which matched-paired measure were collected.</li>
<li>Columns <em>unmatched_condition_x</em> and
<em>unmatched_condition_y</em> uniquely identify the two different sets
of matched-paired measures under analysis.</li>
<li>Columns <em>x.a.i</em>, <em>y.a.i</em>, <em>x.b.i</em> and
<em>y.b.i</em>, where <em>i</em> represents integers to differentiate
each specific matched-pair of measures, contain the actual measures
under analysis.</li>
<li>As many informative columns as needed by users to contextualize the
results of each test. The name of these columns should not contain the
term <em>condition</em> or have the same structure as the
<em>x.a.i</em>, <em>y.a.i</em>, <em>x.b.i</em> and <em>y.b.i</em>
columns. While these columns are optional when running a single test, at
least one column is required when running multiple tests
simultaneously.</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="mbmcw-testing-process">mbMCW testing process<a class="anchor" aria-label="anchor" href="#mbmcw-testing-process"></a>
</h4>
<p>The function <em>mbMCWtest</em> eliminates any matched-paired
measures with at least one missing value (NA) before proceeding with the
following steps.</p>
<ul>
<li>
<p>To estimate the differential bias between the two sets of
matched-paired measures in the dataset, the function <em>mbMCWtest</em>
perfoms the following tasks:</p>
<ul>
<li>For each matched-paired measure, it subtracts the values for the two
possible matched condition contrasts (<em>e.g.</em>, <em>a-b</em> and
<em>b-a</em>).</li>
<li>For each matched condition contrast, it ranks the absolute values of
non-zero differences from lowest to highest. Measure pair differences
with a value of 0 are assigned a 0 rank. If multiple measure pair
differences have the same absolute value, all tied measure pair
differences are assigned the lowest rank possible.</li>
<li>It assigns each measure pair rank a sign based on the sign of its
corresponding measure pair difference.</li>
<li>For each set of matched-paired measures (<em>e.g.</em>, <em>x</em>
and <em>y</em>), it sums the signed ranks for each matched condition
contrast (<em>e.g.</em>, <em>a-b</em> and <em>b-a</em>).</li>
<li>For each set of matched-paired measures (<em>e.g.</em>, <em>x</em>
and <em>y</em>) and each matched condition contrast (<em>e.g.</em>,
<em>a-b</em> and <em>b-a</em>), it calculates one mbMCW_BI. This value
is obtained by dividing each sum of signed ranks by the maximum number
this sum could have if the corresponding measure pairs had the highest
possible positive ranks. Consequently, mbMCW_BI ranges between 1 when
all the values for matched-pair measure differences in the set under
analysis have the highest positive values, and -1 when all the values
for matched-pair measure differences in the set under analysis have the
lowest negative values.</li>
</ul>
</li>
<li>
<p>To assess the significance of the mbMCW_BIs obtained from the
user-provided dataset (observed mbMCW_BIs), the function
<em>mbMCWtest</em> perfoms the following tasks:</p>
<ul>
<li>
<p>It generates a collection of expected-by-chance mbMCW_BIs. These
expected values are obtained by rearranging the matched-pair measures
between the two sets multiple times. The user-provided parameter
<em>max_rearrangements</em> determines the two paths the function
<em>mbMCWtest</em> can follow to generate the collection of
expected-by-chance mbMCW_BIs:</p>
<ul>
<li>
<em>mbMCW exact testing</em>: If the number of distinct
matched-paired measure rearrangements that can alter their initial set
distribution is less than <em>max_rearrangements</em>, the function
<em>mbMCWtest</em> calculates mbMCW_BIs for all possible data
rearrangements.</li>
<li>
<em>mbMCW approximated testing</em>: If the number of distinct
matched-paired measure rearrangements that can alter their initial set
distribution is greater than <em>max_rearrangements</em>, the function
<em>mbMCWtest</em> performs N = <em>max_rearrangements</em> random
measure rearrangements to calculate the collection of expected-by-chance
mbMCW_BIs.</li>
</ul>
</li>
<li><p>It calculates <em>P<sub>upper</sub></em> and
<em>P<sub>lower</sub></em> values as the fraction of expected-by-chance
mbMCW_BIs that are higher or equal to and lower or equal to the observed
mbMCW_BIs, respectively.</p></li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="mbmcw-testing-results">mbMCW testing results<a class="anchor" aria-label="anchor" href="#mbmcw-testing-results"></a>
</h4>
<p>The <em>mbMCWtest</em> function reports to the console the total
number of tests it will execute, and their exact and approximated
counts. It also creates a CSV file named
<em>X_mbMCWtest_results.csv</em>, where <em>X</em> is a user-defined
identifier for the entry dataset CSV file. The
<em>X_mbMCWtest_results.csv</em> file contains four rows for each
mbMCWtest, with mbMCW_BIs calculated for each possible contrast between
matched and unmatched measures (<em>e.g.</em>, <em>a-b</em>,
<em>b-a</em>, <em>x-y</em> and <em>y-x</em>).</p>
<p>The <em>X_mbMCWtest_results.csv</em> file includes the following
columns:</p>
<ul>
<li>User-provided informative columns to contextualize the results of
each test.</li>
<li>Columns <em>matched_condition_a</em> and
<em>matched_condition_b</em> indicate the conditions for which
matched-paired measures were provided.</li>
<li>Columns <em>unmatched_condition_x</em> and
<em>unmatched_condition_y</em> indicate the two sets of matched-pairs
measures.</li>
<li>Columns <em>N</em>, <em>N_x</em> and <em>N_y</em> indicate the total
number of matched-paired measures, and their distribution between the
two unmatched sets after removing any matched-pair with missing values
(NAs).</li>
<li>Column <em>test_type</em> distinguishes between exact and
approximated tests.</li>
<li>Column <em>BI_type</em> indicates mbMCW_BI.</li>
<li>Column <em>matched_condition_contrast</em> and
<em>unmatched_condition_contrast</em> indicate the matched and unmatched
condition contrast for each row of results.</li>
<li>Column <em>observed_BI</em> contains the value of mbMCW_BIs obtained
from analyzing the user-provided dataset.</li>
<li>Column <em>expected_by_chance_BI_N</em> indicates the number of data
rearrangements used to calculate the expected-by-chance mbMCW_BIs. This
value corresponds to the lowest number between all possible measure
rearrangements and the parameter <em>max_rearrangements</em>.</li>
<li>Columns <em>pupper</em> and <em>plower</em> represent the
<em>P<sub>upper</sub></em> and <em>P<sub>lower</sub></em> values,
respectively. They denote the fraction of expected-by-chance mbMCW_BIs
with values higher or equal to and lower or equal to the observed
mbMCW_BIs, respectively.</li>
</ul>
</div>
<div class="section level4">
<h4 id="examples-2">Examples<a class="anchor" aria-label="anchor" href="#examples-2"></a>
</h4>
<div class="section level5">
<h5 id="mbmcw-testing-of-multiple-datasets-organized-vertically">mbMCW testing of multiple datasets organized vertically<a class="anchor" aria-label="anchor" href="#mbmcw-testing-of-multiple-datasets-organized-vertically"></a>
</h5>
<p>This example of mbMCW testing entry dataset with a vertical layout
showcases data for two ideal tests (Table 9). The <em>Contrast</em> I
test includes data for two sets of matched-paired measures randomly
selected from the same range (Figure 5). The <em>Contrast</em> II test
includes data for two sets of matched-paired measures. For the first
set, matched-pair measures were randomly selected from the same range.
For the second set, matched-pair measures were randomly selected from
two different ranges, with the measures corresponding to the second
condition of each matched-pair being generally higher than the measures
corresponding to the first condition of each matched-pair (Figure 5).
The results of these mbMCW tests align with the intended structure of
the entry datasets (Table 10).</p>
<div class="section level6">
<h6 id="table-9--mbmcw-testing-entry-dataset-with-vertical-layout">Table 9. mbMCW testing entry dataset with vertical layout<a class="anchor" aria-label="anchor" href="#table-9--mbmcw-testing-entry-dataset-with-vertical-layout"></a>
</h6>
<pre><code>    contrast matched_condition_a matched_condition_b unmatched_condition
      &lt;char&gt;              &lt;char&gt;              &lt;char&gt;              &lt;char&gt;
 1:        I                AAAA                BBBB                XXXX
 2:        I                AAAA                BBBB                XXXX
 3:        I                AAAA                BBBB                XXXX
 4:        I                AAAA                BBBB                XXXX
 5:        I                AAAA                BBBB                XXXX
 6:        I                AAAA                BBBB                YYYY
 7:        I                AAAA                BBBB                YYYY
 8:        I                AAAA                BBBB                YYYY
 9:        I                AAAA                BBBB                YYYY
10:        I                AAAA                BBBB                YYYY
11:       II                AAAA                BBBB                XXXX
12:       II                AAAA                BBBB                XXXX
13:       II                AAAA                BBBB                XXXX
14:       II                AAAA                BBBB                XXXX
15:       II                AAAA                BBBB                XXXX
16:       II                AAAA                BBBB                YYYY
17:       II                AAAA                BBBB                YYYY
18:       II                AAAA                BBBB                YYYY
19:       II                AAAA                BBBB                YYYY
20:       II                AAAA                BBBB                YYYY
    contrast matched_condition_a matched_condition_b unmatched_condition
    value_a value_b
      &lt;int&gt;   &lt;int&gt;
 1:       6      37
 2:      25      18
 3:      13      27
 4:      48      16
 5:      44      70
 6:      30      19
 7:       9      24
 8:      16      47
 9:      47      33
10:      14      23
11:      13      47
12:      14      41
13:      42      13
14:      40      38
15:      47      11
16:      36      70
17:      38      64
18:      31      60
19:      33      67
20:      25      69
    value_a value_b</code></pre>
</div>
<div class="section level6">
<h6 id="figure-5--data-structure-of-mbmcw-testing-entry-dataset-with-vertical-layout">Figure 5. Data structure of mbMCW testing entry dataset with
vertical layout<a class="anchor" aria-label="anchor" href="#figure-5--data-structure-of-mbmcw-testing-entry-dataset-with-vertical-layout"></a>
</h6>
<p><img src="MCWtests_files/figure-html/unnamed-chunk-16-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="table-10--results-of-mbmcw-testing-using-an-entry-dataset-with-vertical-layout">Table 10. Results of mbMCW testing using an entry dataset with
vertical layout<a class="anchor" aria-label="anchor" href="#table-10--results-of-mbmcw-testing-using-an-entry-dataset-with-vertical-layout"></a>
</h6>
<pre><code>   contrast matched_condition_a matched_condition_b unmatched_condition_x
     &lt;char&gt;              &lt;char&gt;              &lt;char&gt;                &lt;char&gt;
1:        I                AAAA                BBBB                  XXXX
2:        I                AAAA                BBBB                  XXXX
3:        I                AAAA                BBBB                  XXXX
4:        I                AAAA                BBBB                  XXXX
5:       II                AAAA                BBBB                  XXXX
6:       II                AAAA                BBBB                  XXXX
7:       II                AAAA                BBBB                  XXXX
8:       II                AAAA                BBBB                  XXXX
   unmatched_condition_y     N   N_x   N_y    test_type  BI_type
                  &lt;char&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;       &lt;char&gt;   &lt;char&gt;
1:                  YYYY    10     5     5 approximated mbMCW_BI
2:                  YYYY    10     5     5 approximated mbMCW_BI
3:                  YYYY    10     5     5 approximated mbMCW_BI
4:                  YYYY    10     5     5 approximated mbMCW_BI
5:                  YYYY    10     5     5 approximated mbMCW_BI
6:                  YYYY    10     5     5 approximated mbMCW_BI
7:                  YYYY    10     5     5 approximated mbMCW_BI
8:                  YYYY    10     5     5 approximated mbMCW_BI
   matched_condition_contrast unmatched_condition_contrast observed_BI
                       &lt;char&gt;                       &lt;char&gt;       &lt;num&gt;
1:                  AAAA-BBBB                    XXXX-YYYY  0.01886792
2:                  AAAA-BBBB                    YYYY-XXXX -0.01886792
3:                  BBBB-AAAA                    XXXX-YYYY -0.01886792
4:                  BBBB-AAAA                    YYYY-XXXX  0.01886792
5:                  AAAA-BBBB                    XXXX-YYYY  0.64705882
6:                  AAAA-BBBB                    YYYY-XXXX -0.64705882
7:                  BBBB-AAAA                    XXXX-YYYY -0.64705882
8:                  BBBB-AAAA                    YYYY-XXXX  0.64705882
   expected_by_chance_BI_N pupper plower
                     &lt;int&gt;  &lt;num&gt;  &lt;num&gt;
1:                     200  0.470  0.570
2:                     200  0.570  0.470
3:                     200  0.570  0.470
4:                     200  0.470  0.570
5:                     200  0.035  0.980
6:                     200  0.980  0.035
7:                     200  0.980  0.035
8:                     200  0.035  0.980</code></pre>
</div>
</div>
<div class="section level5">
<h5 id="mbmcw-testing-of-multiple-datasets-organized-horizontally">mbMCW testing of multiple datasets organized horizontally<a class="anchor" aria-label="anchor" href="#mbmcw-testing-of-multiple-datasets-organized-horizontally"></a>
</h5>
<p>This example of mbMCW testing entry dataset with a vertical layout
showcases data for six tests (Table 11). It includes two types of
informative columns to accommodate cases where users might want to
perform multiple tests with a nested structure. The columns with the
<em>contrast</em> prefix provide contextual information for groups of
tests or rows, while the columns with the <em>element</em> prefix
provide contextual information for each test or row. Each
<em>Contrast</em> I row includes two sets of matched-paired measures
randomly selected from the same range (Figure 6). Each <em>Contrast</em>
II row includes two sets of matched-paired measures. For the first set,
matched-pair measures were randomly selected from the same range. For
the second set, matched-pair measures were randomly selected from two
different ranges, with the measures corresponding to the second
condition of each matched-pair being generally higher than the measures
corresponding to the first condition of each matched-pair (Figure 6).
The results of these mbMCW tests align with the intended structure of
the entry datasets (Table 12).</p>
<div class="section level6">
<h6 id="table-11--mbmcw-testing-entry-dataset-with-horizontal-layout">Table 11. mbMCW testing entry dataset with horizontal layout<a class="anchor" aria-label="anchor" href="#table-11--mbmcw-testing-entry-dataset-with-horizontal-layout"></a>
</h6>
<pre><code>   contrast contrast_trait element_ID element_chr element_start element_end
     &lt;char&gt;         &lt;char&gt;     &lt;char&gt;       &lt;int&gt;         &lt;int&gt;       &lt;int&gt;
1:        I        trait_a         x1           1          1000        2000
2:        I        trait_a         x2           1          5000        5500
3:        I        trait_a         x3           1         90000      100000
4:       II        trait_b         x1           1          1000        2000
5:       II        trait_b         x2           1          5000        5500
6:       II        trait_b         x3           1         90000      100000
   matched_condition_a matched_condition_b unmatched_condition_x
                &lt;char&gt;              &lt;char&gt;                &lt;char&gt;
1:                AAAA                BBBB                  XXXX
2:                AAAA                BBBB                  XXXX
3:                AAAA                BBBB                  XXXX
4:                AAAA                BBBB                  XXXX
5:                AAAA                BBBB                  XXXX
6:                AAAA                BBBB                  XXXX
   unmatched_condition_y x.a.1 x.a.2 x.a.3 x.a.4 x.a.5 y.a.6 y.a.7 y.a.8 y.a.9
                  &lt;char&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
1:                  YYYY     3    31    21     4    44    27    48    17    36
2:                  YYYY    20     7    30    27    40    34    34    17     7
3:                  YYYY    14    12    12    32    10    27    36    42    10
4:                  YYYY    12    16    35    36    38    36    49     6    22
5:                  YYYY     7    16    28    20    17    42    14    20    43
6:                  YYYY     5    30    17    46    22    47    13    32    20
   y.a.10 x.b.1 x.b.2 x.b.3 x.b.4 x.b.5 y.b.6 y.b.7 y.b.8 y.b.9 y.b.10
    &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;  &lt;int&gt;
1:      7    42    70    17    30    17    18    36    20    36     48
2:     26    41    19     2     1    48    22    34     8    11      7
3:      5    25    27    23    13    14     2    43    35    21     23
4:      2    15    18    17    33    14    68    64    77    75     63
5:     29    26    31    25    48    16    54    61    51    68     68
6:     28    19    42    26    26    13    78    64    70    51     52</code></pre>
</div>
<div class="section level6">
<h6 id="figure-6--data-structure-of-mbmcw-testing-entry-dataset-with-horizontal-layout">Figure 6. Data structure of mbMCW testing entry dataset with
horizontal layout<a class="anchor" aria-label="anchor" href="#figure-6--data-structure-of-mbmcw-testing-entry-dataset-with-horizontal-layout"></a>
</h6>
<p><img src="MCWtests_files/figure-html/unnamed-chunk-19-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="table-12--results-of-mbmcw-testing-using-an-entry-dataset-with-horizontal-layout">Table 12. Results of mbMCW testing using an entry dataset with
horizontal layout<a class="anchor" aria-label="anchor" href="#table-12--results-of-mbmcw-testing-using-an-entry-dataset-with-horizontal-layout"></a>
</h6>
<pre><code>    contrast contrast_trait element_ID element_chr element_start element_end
      &lt;char&gt;         &lt;char&gt;     &lt;char&gt;       &lt;int&gt;         &lt;int&gt;       &lt;int&gt;
 1:        I        trait_a         x1           1          1000        2000
 2:        I        trait_a         x1           1          1000        2000
 3:        I        trait_a         x1           1          1000        2000
 4:        I        trait_a         x1           1          1000        2000
 5:        I        trait_a         x2           1          5000        5500
 6:        I        trait_a         x2           1          5000        5500
 7:        I        trait_a         x2           1          5000        5500
 8:        I        trait_a         x2           1          5000        5500
 9:        I        trait_a         x3           1         90000      100000
10:        I        trait_a         x3           1         90000      100000
11:        I        trait_a         x3           1         90000      100000
12:        I        trait_a         x3           1         90000      100000
13:       II        trait_b         x1           1          1000        2000
14:       II        trait_b         x1           1          1000        2000
15:       II        trait_b         x1           1          1000        2000
16:       II        trait_b         x1           1          1000        2000
17:       II        trait_b         x2           1          5000        5500
18:       II        trait_b         x2           1          5000        5500
19:       II        trait_b         x2           1          5000        5500
20:       II        trait_b         x2           1          5000        5500
21:       II        trait_b         x3           1         90000      100000
22:       II        trait_b         x3           1         90000      100000
23:       II        trait_b         x3           1         90000      100000
24:       II        trait_b         x3           1         90000      100000
    contrast contrast_trait element_ID element_chr element_start element_end
    matched_condition_a matched_condition_b unmatched_condition_x
                 &lt;char&gt;              &lt;char&gt;                &lt;char&gt;
 1:                AAAA                BBBB                  XXXX
 2:                AAAA                BBBB                  XXXX
 3:                AAAA                BBBB                  XXXX
 4:                AAAA                BBBB                  XXXX
 5:                AAAA                BBBB                  XXXX
 6:                AAAA                BBBB                  XXXX
 7:                AAAA                BBBB                  XXXX
 8:                AAAA                BBBB                  XXXX
 9:                AAAA                BBBB                  XXXX
10:                AAAA                BBBB                  XXXX
11:                AAAA                BBBB                  XXXX
12:                AAAA                BBBB                  XXXX
13:                AAAA                BBBB                  XXXX
14:                AAAA                BBBB                  XXXX
15:                AAAA                BBBB                  XXXX
16:                AAAA                BBBB                  XXXX
17:                AAAA                BBBB                  XXXX
18:                AAAA                BBBB                  XXXX
19:                AAAA                BBBB                  XXXX
20:                AAAA                BBBB                  XXXX
21:                AAAA                BBBB                  XXXX
22:                AAAA                BBBB                  XXXX
23:                AAAA                BBBB                  XXXX
24:                AAAA                BBBB                  XXXX
    matched_condition_a matched_condition_b unmatched_condition_x
    unmatched_condition_y     N   N_x   N_y    test_type  BI_type
                   &lt;char&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;       &lt;char&gt;   &lt;char&gt;
 1:                  YYYY    10     5     5 approximated mbMCW_BI
 2:                  YYYY    10     5     5 approximated mbMCW_BI
 3:                  YYYY    10     5     5 approximated mbMCW_BI
 4:                  YYYY    10     5     5 approximated mbMCW_BI
 5:                  YYYY    10     5     5 approximated mbMCW_BI
 6:                  YYYY    10     5     5 approximated mbMCW_BI
 7:                  YYYY    10     5     5 approximated mbMCW_BI
 8:                  YYYY    10     5     5 approximated mbMCW_BI
 9:                  YYYY    10     5     5 approximated mbMCW_BI
10:                  YYYY    10     5     5 approximated mbMCW_BI
11:                  YYYY    10     5     5 approximated mbMCW_BI
12:                  YYYY    10     5     5 approximated mbMCW_BI
13:                  YYYY    10     5     5 approximated mbMCW_BI
14:                  YYYY    10     5     5 approximated mbMCW_BI
15:                  YYYY    10     5     5 approximated mbMCW_BI
16:                  YYYY    10     5     5 approximated mbMCW_BI
17:                  YYYY    10     5     5 approximated mbMCW_BI
18:                  YYYY    10     5     5 approximated mbMCW_BI
19:                  YYYY    10     5     5 approximated mbMCW_BI
20:                  YYYY    10     5     5 approximated mbMCW_BI
21:                  YYYY    10     5     5 approximated mbMCW_BI
22:                  YYYY    10     5     5 approximated mbMCW_BI
23:                  YYYY    10     5     5 approximated mbMCW_BI
24:                  YYYY    10     5     5 approximated mbMCW_BI
    unmatched_condition_y     N   N_x   N_y    test_type  BI_type
    matched_condition_contrast unmatched_condition_contrast observed_BI
                        &lt;char&gt;                       &lt;char&gt;       &lt;num&gt;
 1:                  AAAA-BBBB                    XXXX-YYYY -0.18181818
 2:                  AAAA-BBBB                    YYYY-XXXX  0.18181818
 3:                  BBBB-AAAA                    XXXX-YYYY  0.18181818
 4:                  BBBB-AAAA                    YYYY-XXXX -0.18181818
 5:                  AAAA-BBBB                    XXXX-YYYY -0.18181818
 6:                  AAAA-BBBB                    YYYY-XXXX  0.18181818
 7:                  BBBB-AAAA                    XXXX-YYYY  0.18181818
 8:                  BBBB-AAAA                    YYYY-XXXX -0.18181818
 9:                  AAAA-BBBB                    XXXX-YYYY -0.09803922
10:                  AAAA-BBBB                    YYYY-XXXX  0.09803922
11:                  BBBB-AAAA                    XXXX-YYYY  0.09803922
12:                  BBBB-AAAA                    YYYY-XXXX -0.09803922
13:                  AAAA-BBBB                    XXXX-YYYY  0.88888889
14:                  AAAA-BBBB                    YYYY-XXXX -0.88888889
15:                  BBBB-AAAA                    XXXX-YYYY -0.88888889
16:                  BBBB-AAAA                    YYYY-XXXX  0.88888889
17:                  AAAA-BBBB                    XXXX-YYYY  0.41818182
18:                  AAAA-BBBB                    YYYY-XXXX -0.41818182
19:                  BBBB-AAAA                    XXXX-YYYY -0.41818182
20:                  BBBB-AAAA                    YYYY-XXXX  0.41818182
21:                  AAAA-BBBB                    XXXX-YYYY  0.69811321
22:                  AAAA-BBBB                    YYYY-XXXX -0.69811321
23:                  BBBB-AAAA                    XXXX-YYYY -0.69811321
24:                  BBBB-AAAA                    YYYY-XXXX  0.69811321
    matched_condition_contrast unmatched_condition_contrast observed_BI
    expected_by_chance_BI_N pupper plower
                      &lt;int&gt;  &lt;num&gt;  &lt;num&gt;
 1:                     200  0.660  0.375
 2:                     200  0.375  0.660
 3:                     200  0.375  0.660
 4:                     200  0.660  0.375
 5:                     200  0.720  0.315
 6:                     200  0.315  0.720
 7:                     200  0.315  0.720
 8:                     200  0.720  0.315
 9:                     200  0.630  0.425
10:                     200  0.425  0.630
11:                     200  0.425  0.630
12:                     200  0.630  0.425
13:                     200  0.000  1.000
14:                     200  1.000  0.000
15:                     200  1.000  0.000
16:                     200  0.000  1.000
17:                     200  0.035  0.975
18:                     200  0.975  0.035
19:                     200  0.975  0.035
20:                     200  0.035  0.975
21:                     200  0.005  1.000
22:                     200  1.000  0.005
23:                     200  1.000  0.005
24:                     200  0.005  1.000
    expected_by_chance_BI_N pupper plower</code></pre>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="bias-measures-mcw-bmcw-test">4.4 bias-measures MCW (bMCW) test<a class="anchor" aria-label="anchor" href="#bias-measures-mcw-bmcw-test"></a>
</h3>
<div class="section level4">
<h4 id="introduction-3">Introduction<a class="anchor" aria-label="anchor" href="#introduction-3"></a>
</h4>
<p>The bMCW test is actually a combination of two tests that assess
whether a set of measures of bias for a quantitative trait between two
conditions or a subset of these bias measures are themselves
significantly biased in the same direction. For instance, bMCW tests can
be used to analyze bias indexes obtained using other MCW tests or fold
change for transcript abundances spanning the entire transcriptome or
only for genes located in specific genomic regions from two sets of mice
exposed to different conditions.</p>
</div>
<div class="section level4">
<h4 id="bmcw-testing-entry-dataset-formatting">bMCW testing entry dataset formatting<a class="anchor" aria-label="anchor" href="#bmcw-testing-entry-dataset-formatting"></a>
</h4>
<p>When executing the <em>bMCWtest</em> function, users must provide the
path to a local CSV file named <em>X_bMCWtest_data.csv</em>, where
<em>X</em> serves as a user-defined identifier.
<em>X_bMCWtest_data.csv</em> should include the following columns:</p>
<ul>
<li>Column <em>bias_value</em> contains the value of the bias measure
under analysis.</li>
<li>Columns <em>subset_x</em>, where <em>x</em> represents the specific
type of subset for each column, such as “chr” for chromosomes or “GO”
for Gene Ontology. These columns are required if users intend to assess
whether bias measures for certain subsets of elements in the dataset are
significantly biased in the same direction. Columns <em>subset_x</em>
can indicate whether an element belongs to a subset using either “YES”
and “NO”, or specific subset names like “chr1” or “chrX”, or a
combination of both, such as “chr1”, “chrX” and “NO”. The function
<em>bMCW test</em> will transform the dataset to conduct independent
analysis of each subset of elements marked as “YES” or with a specific
subset name in each <em>subset_x</em> column.</li>
<li>As many informative columns as needed by users to contextualize the
results of each test. The names of these columns should not contain the
terms <em>bias_value</em> or <em>subset</em>. While these columns are
optional when running a single test, at least one column is required
when running multiple tests simultaneously. All rows for each individual
test must contain the same information in these columns.</li>
<li>Users can specify columns with information relevant about each
element or row using the column name structure <em>element_x</em>, where
<em>x</em> indicates the specific information in each column (see
example). However, <em>element_x</em> columns are not essential for bMCW
testing and will not be included in the results file.</li>
</ul>
</div>
<div class="section level4">
<h4 id="bmcw-testing-process">bMCW testing process<a class="anchor" aria-label="anchor" href="#bmcw-testing-process"></a>
</h4>
<p>The function <em>bMCWtest</em> eliminates missing values (NAs) from
the dataset before proceeding with the following steps.</p>
<ul>
<li>
<p>To estimate the bias for all bias measures in the entire dataset
or a subset of them, the function <em>bMCWtest</em> performs the
following tasks:</p>
<ul>
<li>It ranks all bias measures with non-zero values from lowest to
highest. Bias measures with a value of 0 are assigned a 0 rank. If
multiple bias measures have the same absolute value, all tied bias
measures are asssigned the lowest rank possible.</li>
<li>It assigns each rank a sign based on the sign of its corresponding
bias measure.</li>
<li>It calculates a whole-set bias index (bMCW_wBI) by summing the
signed ranks for all elements in the dataset and dividing it by the
maximum number that sum could have if all bias measures were positive.
Consequently, bMCW_wBI ranges between 1 when all bias measures are
positive, and -1 when all bias measures are negative.</li>
<li>It calculates a subset bias index (bMCW_sBI) for each subset of
elements under analysis by summing the signed ranks for the elements in
the subset and dividing it by the maximum number that sum could have if
the elements in the subset had the highest possible positive bias
measures. Consequently, bMCW_sBI ranges between 1 when the bias measures
for the subset in question have the highest positive bias measures in
the entire dataset, and -1 when the bias measures for the subset in
question have the lowest negative bias measures in the entire
dataset.</li>
</ul>
</li>
<li><p>To assess the significance of the bMCW-wBIs and bMCW-sBIs
obtained from the user-provided dataset (observed bMCW-wBIs and
bMCW-sBIs), the function <em>bMCWtest</em> performs the following
tasks,</p></li>
</ul>
<pre><code>- It generates a collection of expected-by-chance bMCW_wBIs by rearranging the signs of all signed ranks multiple times. The function *bMCWtest* also generates a collection of expected-by-chance bMCW_sBIs by rearranging the subset of elements multiple times. The user-provided parameter *max_rearrangements* determines the two paths that the function *bMCWtest* can follow to generate the collection of expected-by-chance bMCW_wBIs and bMCW_sBIs:

    - *bMCW exact testing*: If the number of distinct bias measure rearrangements that can alter their initial sign distribution or subset distribution is less than *max_rearrangements*, the function *bMCWtest* calculates bMCW_wBIs or bMCW_sBIs for all possible data rearrangements.
    - *bMCW approximated testing*: If the number of distinct bias measure rearrangements that can alter their initial sign distribution or subset distribution is greater than *max_rearrangements*, the function *bMCWtest* performs N = *max_rearrangements* random measure rearrangements to calculate the collection of expected-by-chance bMCW_wBIs or bMCW_sBIs.

- It calculates *P~upper~* and *P~lower~* values, as the fraction of expected-by-chance bMCW-wBIs and bMCW-sBIs that are higher or equal to and lower or equal to the observed bMCW-wBIs and bMCW-sBIs, respectively.</code></pre>
</div>
<div class="section level4">
<h4 id="bmcw-testing-results">bMCW testing results<a class="anchor" aria-label="anchor" href="#bmcw-testing-results"></a>
</h4>
<p>The <em>bMCWtest</em> function reports to the console the total
number of tests it will execute, and their exact and approximated
counts. It also creates a CSV file named
<em>X_bMCWtest_results.csv</em>, where <em>X</em> is a user-defined
identifier for the entry dataset CSV file. The
<em>X_bMCWtest_results.csv</em> file contains one row for each bMCWtest
to indicate the results of whole-set bMCW testing, and as many rows as
necessary to indicate the results of subset bMCW testing. Rows for
whole-set analyses will be at the top of <em>X_bMCWtest_results.csv</em>
file.</p>
<p>The <em>X_bMCWtest_results.csv</em> file includes the following
columns:</p>
<ul>
<li>User-provided informative columns to contextualize the results of
each test.</li>
<li>Column <em>subset_type</em> indicates whether the results in each
row corresponds to <em>whole-set</em> tests or specific <em>subset</em>
tests, such as “chr” for chromosomes or “GO” for Gene Ontology
terms.</li>
<li>Column <em>tested_subset</em> indicates the name of the subset under
analysis. For <em>whole-set</em> tests, the <em>tested_subset</em>
column indicates “none”. For subset tests, the <em>tested_subset</em>
column indicates “YES” or the specific name of the subset under
analysis, such as “chr1” or “chrX”.</li>
<li>Columns <em>N</em> and <em>n</em> indicate the total number of
elements in the whole set and those associated with the subset under
analysis, respectively, after removing missing values (NAs). For
<em>whole-set</em> tests, columns <em>N</em> and <em>n</em> have the
same value.</li>
<li>Column <em>test_type</em> distinguishes between exact and
approximated tests.</li>
<li>Column <em>BI_type</em> indicates whether results correspond to
<em>whole-set</em> tests (bMCW_wBI) or to <em>subset</em> tests
(bMCW_sBIs).</li>
<li>Column <em>observed_BI</em> contains the value of bMCW_BIs obtained
from analyzing the user-provided dataset.</li>
<li>Column <em>expected_by_chance_BI_N</em> indicates the number of data
rearrangements used to calculate the expected-by-chance bMCW_wBIs and
bMCW_sBIs. This value corresponds to the lowest number between all
possible measure rearrangements and the parameter
<em>max_rearrangements</em>.</li>
<li>Columns <em>pupper</em> and <em>plower</em> represent the
<em>P<sub>upper</sub></em> and <em>P<sub>lower</sub></em> values,
respectively. They denote the fraction of expected-by-chance bMCW_wBIs
or bMCW_sBIs with values higher or equal to and lower or equal to the
observed bMCW_wBIs or bMCW_sBIs, respectively.</li>
</ul>
</div>
<div class="section level4">
<h4 id="examples-3">Examples<a class="anchor" aria-label="anchor" href="#examples-3"></a>
</h4>
<div class="section level5">
<h5 id="bmcw-testing-of-multiple-datasets">bMCW testing of multiple datasets<a class="anchor" aria-label="anchor" href="#bmcw-testing-of-multiple-datasets"></a>
</h5>
<p>This example of bMCW testing entry dataset includes bias measures for
three tests, each with ten elements distributed across four subsets:
chr1, chr2, chrX and chrY (Table 13), In this example, the bias measures
under study are uMCW_BIs, which range from 1 and -1. The
<em>Contrast</em> I test includes bias measures randomly selected from
the same range (Figure 7). The <em>Contrast</em> II test includes bias
measures randomly selected from the same range, with most of them having
positive values (Figure 7). The <em>Contrast</em> III includes bias
measures randomly selected from the same range for three of the four
subsets, while bias measures for the fourth subset are randomly selected
with the largest positive values in the dataset (Figure 7). The results
of these bMCW tests align with the intended structure of the entry
datasets (Table 14).</p>
<div class="section level6">
<h6 id="table-13--bmcw-testing-entry-dataset">Table 13. bMCW testing entry dataset<a class="anchor" aria-label="anchor" href="#table-13--bmcw-testing-entry-dataset"></a>
</h6>
<pre><code>    contrast contrast_trait element_ID element_chr element_start element_end
      &lt;char&gt;         &lt;char&gt;     &lt;char&gt;      &lt;char&gt;         &lt;int&gt;       &lt;int&gt;
 1:        I        trait_a         x1        chr1          1000        2000
 2:        I        trait_a         x2        chr1          5000        5500
 3:        I        trait_a         x3        chr1         90000      100000
 4:        I        trait_a         x4        chr2           150         300
 5:        I        trait_a         x5        chr2          2545        7000
 6:        I        trait_a         x6        chr2         80000      100000
 7:        I        trait_a         x7        chrX          4000        7000
 8:        I        trait_a         x8        chrX          9000       10000
 9:        I        trait_a         x9        chrX         30000       31000
10:        I        trait_a        x10        chrY           800        1000
11:       II        trait_b         x1        chr1          1000        2000
12:       II        trait_b         x2        chr1          5000        5500
13:       II        trait_b         x3        chr1         90000      100000
14:       II        trait_b         x4        chr2           150         300
15:       II        trait_b         x5        chr2          2545        7000
16:       II        trait_b         x6        chr2         80000      100000
17:       II        trait_b         x7        chrX          4000        7000
18:       II        trait_b         x8        chrX          9000       10000
19:       II        trait_b         x9        chrX         30000       31000
20:       II        trait_b        x10        chrY           800        1000
21:      III        trait_c         x1        chr1          1000        2000
22:      III        trait_c         x2        chr1          5000        5500
23:      III        trait_c         x3        chr1         90000      100000
24:      III        trait_c         x4        chr2           150         300
25:      III        trait_c         x5        chr2          2545        7000
26:      III        trait_c         x6        chr2         80000      100000
27:      III        trait_c         x7        chrX          4000        7000
28:      III        trait_c         x8        chrX          9000       10000
29:      III        trait_c         x9        chrX         30000       31000
30:      III        trait_c        x10        chrY           800        1000
    contrast contrast_trait element_ID element_chr element_start element_end
    subset_chr bias_measure condition_contrast  bias_value
        &lt;char&gt;       &lt;char&gt;             &lt;char&gt;       &lt;num&gt;
 1:       chr1      uMCW_BI          AAAA-BBBB  0.59883185
 2:       chr1      uMCW_BI          AAAA-BBBB -0.37376010
 3:       chr1      uMCW_BI          AAAA-BBBB  0.75433521
 4:       chr2      uMCW_BI          AAAA-BBBB -0.36807420
 5:       chr2      uMCW_BI          AAAA-BBBB -0.97818020
 6:       chr2      uMCW_BI          AAAA-BBBB  0.59299509
 7:       chrX      uMCW_BI          AAAA-BBBB -0.58644011
 8:       chrX      uMCW_BI          AAAA-BBBB -0.36797886
 9:       chrX      uMCW_BI          AAAA-BBBB -0.63406711
10:       chrY      uMCW_BI          AAAA-BBBB  0.17762437
11:       chr1      uMCW_BI          AAAA-BBBB -0.03649558
12:       chr1      uMCW_BI          AAAA-BBBB  0.74867686
13:       chr1      uMCW_BI          AAAA-BBBB  0.81900144
14:       chr2      uMCW_BI          AAAA-BBBB  0.95044745
15:       chr2      uMCW_BI          AAAA-BBBB  0.87731860
16:       chr2      uMCW_BI          AAAA-BBBB  0.30838414
17:       chrX      uMCW_BI          AAAA-BBBB  0.44241725
18:       chrX      uMCW_BI          AAAA-BBBB  0.83965786
19:       chrX      uMCW_BI          AAAA-BBBB  0.65278713
20:       chrY      uMCW_BI          AAAA-BBBB -0.41912458
21:       chr1      uMCW_BI          AAAA-BBBB  0.42249472
22:       chr1      uMCW_BI          AAAA-BBBB  0.69599878
23:       chr1      uMCW_BI          AAAA-BBBB  0.59133980
24:       chr2      uMCW_BI          AAAA-BBBB -0.04787103
25:       chr2      uMCW_BI          AAAA-BBBB  0.01716967
26:       chr2      uMCW_BI          AAAA-BBBB  0.09079163
27:       chrX      uMCW_BI          AAAA-BBBB -0.07600623
28:       chrX      uMCW_BI          AAAA-BBBB -0.09808774
29:       chrX      uMCW_BI          AAAA-BBBB -0.02192925
30:       chrY      uMCW_BI          AAAA-BBBB -0.08190336
    subset_chr bias_measure condition_contrast  bias_value</code></pre>
</div>
<div class="section level6">
<h6 id="figure-7--data-structure-of-bmcw-testing-entry-dataset">Figure 7. Data structure of bMCW testing entry dataset<a class="anchor" aria-label="anchor" href="#figure-7--data-structure-of-bmcw-testing-entry-dataset"></a>
</h6>
<p><img src="MCWtests_files/figure-html/unnamed-chunk-22-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="table-14--bmcw-testing-results">Table 14. bMCW testing results<a class="anchor" aria-label="anchor" href="#table-14--bmcw-testing-results"></a>
</h6>
<pre><code>    contrast contrast_trait bias_measure condition_contrast subset_type
      &lt;char&gt;         &lt;char&gt;       &lt;char&gt;             &lt;char&gt;      &lt;char&gt;
 1:        I        trait_a      uMCW_BI          AAAA-BBBB    wholeset
 2:       II        trait_b      uMCW_BI          AAAA-BBBB    wholeset
 3:      III        trait_c      uMCW_BI          AAAA-BBBB    wholeset
 4:        I        trait_a      uMCW_BI          AAAA-BBBB  subset_chr
 5:        I        trait_a      uMCW_BI          AAAA-BBBB  subset_chr
 6:        I        trait_a      uMCW_BI          AAAA-BBBB  subset_chr
 7:        I        trait_a      uMCW_BI          AAAA-BBBB  subset_chr
 8:       II        trait_b      uMCW_BI          AAAA-BBBB  subset_chr
 9:       II        trait_b      uMCW_BI          AAAA-BBBB  subset_chr
10:       II        trait_b      uMCW_BI          AAAA-BBBB  subset_chr
11:       II        trait_b      uMCW_BI          AAAA-BBBB  subset_chr
12:      III        trait_c      uMCW_BI          AAAA-BBBB  subset_chr
13:      III        trait_c      uMCW_BI          AAAA-BBBB  subset_chr
14:      III        trait_c      uMCW_BI          AAAA-BBBB  subset_chr
15:      III        trait_c      uMCW_BI          AAAA-BBBB  subset_chr
    tested_subset     N     n    test_type  BI_type observed_BI
           &lt;char&gt; &lt;int&gt; &lt;int&gt;       &lt;char&gt;   &lt;char&gt;       &lt;num&gt;
 1:          none    10    10 approximated bMCW_wBI  -0.1636364
 2:          none    10    10 approximated bMCW_wBI   0.8545455
 3:          none    10    10 approximated bMCW_wBI   0.2363636
 4:          chr1    10     3        exact bMCW_sBI   0.4444444
 5:          chr2    10     3        exact bMCW_sBI  -0.2592593
 6:          chrX    10     3        exact bMCW_sBI  -0.5555556
 7:          chrY    10     1        exact bMCW_sBI   0.1000000
 8:          chr1    10     3        exact bMCW_sBI   0.4444444
 9:          chr2    10     3        exact bMCW_sBI   0.7777778
10:          chrX    10     3        exact bMCW_sBI   0.6296296
11:          chrY    10     1        exact bMCW_sBI  -0.3000000
12:          chr1    10     3        exact bMCW_sBI   1.0000000
13:          chr2    10     3        exact bMCW_sBI   0.1481481
14:          chrX    10     3        exact bMCW_sBI  -0.4814815
15:          chrY    10     1        exact bMCW_sBI  -0.5000000
    expected_by_chance_BI_N      pupper     plower
                      &lt;int&gt;       &lt;num&gt;      &lt;num&gt;
 1:                     200 0.435000000 0.64500000
 2:                     200 0.050000000 0.96500000
 3:                     200 0.090000000 0.92500000
 4:                     120 0.075000000 0.94166667
 5:                     120 0.700000000 0.33333333
 6:                     120 0.900000000 0.12500000
 7:                      10 0.400000000 0.70000000
 8:                     120 0.658333333 0.40000000
 9:                     120 0.175000000 0.87500000
10:                     120 0.366666667 0.68333333
11:                      10 1.000000000 0.10000000
12:                     120 0.008333333 1.00000000
13:                     120 0.483333333 0.55833333
14:                     120 0.966666667 0.04166667
15:                      10 0.900000000 0.20000000</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body" entry-spacing="0" line-spacing="2">
<div id="ref-díaz-castillo2013" class="csl-entry">
<div class="csl-left-margin">1. </div>
<div class="csl-right-inline">Díaz-Castillo, C. <a href="https://doi.org/10.1371/journal.pone.0064491" class="external-link">Females and Males
Contribute in Opposite Ways to the Evolution of Gene Order in
Drosophila</a>. <em>PLoS ONE</em> <strong>8</strong>, e64491
(2013).</div>
</div>
<div id="ref-díaz-castillo2017" class="csl-entry">
<div class="csl-left-margin">2. </div>
<div class="csl-right-inline">Díaz-Castillo, C. <a href="https://doi.org/10.7717/peerj.4004" class="external-link">Transcriptome dynamics along
axolotl regenerative development are consistent with an extensive
reduction in gene expression heterogeneity in dedifferentiated
cells</a>. <em>PeerJ</em> <strong>5</strong>, e4004 (2017).</div>
</div>
<div id="ref-chamorro-garcia2017" class="csl-entry">
<div class="csl-left-margin">3. </div>
<div class="csl-right-inline">Chamorro-Garcia, R. <em>et al.</em> <a href="https://doi.org/10.1038/s41467-017-01944-z" class="external-link">Ancestral perinatal
obesogen exposure results in a transgenerational thrifty phenotype in
mice</a>. <em>Nature Communications</em> <strong>8</strong>,
(2017).</div>
</div>
<div id="ref-díaz-castillo2018" class="csl-entry">
<div class="csl-left-margin">4. </div>
<div class="csl-right-inline">Díaz-Castillo, C. <a href="https://doi.org/10.1007/s11692-018-9454-y" class="external-link">Same-Sex Twin Pair
Phenotypic Correlations are Consistent with Human Y Chromosome Promoting
Phenotypic Heterogeneity</a>. <em>Evolutionary Biology</em>
<strong>45</strong>, 248–258 (2018).</div>
</div>
<div id="ref-diaz-castillo2019" class="csl-entry">
<div class="csl-left-margin">5. </div>
<div class="csl-right-inline">Diaz-Castillo, C., Chamorro-Garcia, R., Shioda,
T. &amp; Blumberg, B. <a href="https://doi.org/10.1038/s41598-019-49440-2" class="external-link">Transgenerational
Self-Reconstruction of Disrupted Chromatin Organization After Exposure
To An Environmental Stressor in Mice</a>. <em>Scientific Reports</em>
<strong>9</strong>, (2019).</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Carlos Diaz-Castillo, Raquel Chamorro-Garcia.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
